{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "87d9970d-0d30-4488-b896-086215563a8f",
   "metadata": {},
   "source": [
    "# 1. Downloading and assembling microbial sequence data\n",
    "\n",
    "Welcome to the **Downloading and assembling microbial sequence data** supplementary materials. This document is written as a [Jupyter](https://jupyter.org/) notebook, which means that it contains both written documentation as well as runnable code in the same document.\n",
    "\n",
    "There are two methods you can use to run the code here: (1) either copy-paste the bash commands to a separate terminal, or (2) you can run directly in Juptyer."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b48d0acf-c235-4087-9834-8ebd2cbc51e1",
   "metadata": {},
   "source": [
    "## 1.1. Copy-paste commands to a separate terminal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe1a4579-8c0d-4ff7-a7d4-3cc4dc00756b",
   "metadata": {},
   "source": [
    "## 1.2. Running from within Jupyter\n",
    "\n",
    "The easiest way to run the code is to click the **Play** button in the Jupyter notebook as shown below:\n",
    "\n",
    "![run-jupyter.png](../images/run-jupyter.png)\n",
    "\n",
    "This will run a single cell (block of code or text) and advance to the next cell. A cell that's running will be marked with an asterisk `*`:\n",
    "\n",
    "![jupyter-cell-run.png](../images/jupyter-cell-run.png)\n",
    "\n",
    "If you, instead, which to run everything in this notebook all at once you can find the **Kerenel** option in the menu at the top and select **Restart Kerenel and Run All Cells..**, which will run everything in this notebook all at once.\n",
    "\n",
    "![restart-kernel-jupyter.png](../images/restart-kernel-jupyter.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2160c6a-6691-44ad-a8d4-ad0dc8322d14",
   "metadata": {},
   "source": [
    "# 2. Download genomes from NCBI\n",
    "\n",
    "In this section we will go over how to download genomes from [NCBI's Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra/) by using the [SRA Tools](https://github.com/ncbi/sra-tools) command-line software provided by NCBI."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccdc0ad5-173a-4271-9a24-513565b06c35",
   "metadata": {},
   "source": [
    "## 2.1. Install `sra-tools` using conda\n",
    "\n",
    "This will create a new conda environment `sra-tools` and install the package `sra-tools` in this environment.\n",
    "\n",
    "The `-y` option here means to automatically answer `yes` to any prompts for input by conda. This is important when running using Jupyter (but can be left out if you are running via the command-line)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8c678e8d-aa50-4f3b-b393-92026b899c2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/sra-tools\n",
      "\n",
      "  added / updated specs:\n",
      "    - sra-tools\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
      "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
      "  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0\n",
      "  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0\n",
      "  curl               conda-forge/linux-64::curl-7.81.0-h2574ce0_0\n",
      "  hdf5               conda-forge/linux-64::hdf5-1.10.6-nompi_h6a2412b_1114\n",
      "  icu                conda-forge/linux-64::icu-69.1-h9c3ff4c_0\n",
      "  krb5               conda-forge/linux-64::krb5-1.19.2-hcc1bbae_3\n",
      "  libcurl            conda-forge/linux-64::libcurl-7.81.0-h2574ce0_0\n",
      "  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2\n",
      "  libev              conda-forge/linux-64::libev-4.33-h516909a_1\n",
      "  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11\n",
      "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-11.2.0-h69a702a_11\n",
      "  libgfortran5       conda-forge/linux-64::libgfortran5-11.2.0-h5c6108e_11\n",
      "  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11\n",
      "  libiconv           conda-forge/linux-64::libiconv-1.16-h516909a_0\n",
      "  libnghttp2         conda-forge/linux-64::libnghttp2-1.43.0-h812cca2_1\n",
      "  libssh2            conda-forge/linux-64::libssh2-1.10.0-ha56f1ee_2\n",
      "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_11\n",
      "  libxml2            conda-forge/linux-64::libxml2-2.9.12-h885dcf4_1\n",
      "  libzlib            conda-forge/linux-64::libzlib-1.2.11-h36c2ea0_1013\n",
      "  ncbi-ngs-sdk       bioconda/linux-64::ncbi-ngs-sdk-2.11.2-hff44eed_0\n",
      "  ncurses            conda-forge/linux-64::ncurses-6.3-h9c3ff4c_0\n",
      "  openssl            conda-forge/linux-64::openssl-1.1.1l-h7f98852_0\n",
      "  ossuuid            conda-forge/linux-64::ossuuid-1.6.2-hf484d3e_1000\n",
      "  perl               conda-forge/linux-64::perl-5.26.2-h36c2ea0_1008\n",
      "  perl-app-cpanminus bioconda/linux-64::perl-app-cpanminus-1.7044-pl526_1\n",
      "  perl-business-isbn bioconda/linux-64::perl-business-isbn-3.004-pl526_0\n",
      "  perl-business-isb~ bioconda/linux-64::perl-business-isbn-data-20140910.003-pl526_0\n",
      "  perl-carp          bioconda/linux-64::perl-carp-1.38-pl526_3\n",
      "  perl-constant      bioconda/linux-64::perl-constant-1.33-pl526_1\n",
      "  perl-data-dumper   bioconda/linux-64::perl-data-dumper-2.173-pl526_0\n",
      "  perl-encode        bioconda/linux-64::perl-encode-2.88-pl526_1\n",
      "  perl-exporter      bioconda/linux-64::perl-exporter-5.72-pl526_1\n",
      "  perl-extutils-mak~ bioconda/linux-64::perl-extutils-makemaker-7.36-pl526_1\n",
      "  perl-file-path     bioconda/linux-64::perl-file-path-2.16-pl526_0\n",
      "  perl-file-temp     bioconda/linux-64::perl-file-temp-0.2304-pl526_2\n",
      "  perl-mime-base64   bioconda/linux-64::perl-mime-base64-3.15-pl526_1\n",
      "  perl-parent        bioconda/linux-64::perl-parent-0.236-pl526_1\n",
      "  perl-uri           bioconda/linux-64::perl-uri-1.76-pl526_0\n",
      "  perl-xml-libxml    bioconda/linux-64::perl-xml-libxml-2.0132-pl526h7ec2d77_1\n",
      "  perl-xml-namespac~ bioconda/linux-64::perl-xml-namespacesupport-1.12-pl526_0\n",
      "  perl-xml-sax       bioconda/noarch::perl-xml-sax-1.02-pl526_0\n",
      "  perl-xml-sax-base  bioconda/linux-64::perl-xml-sax-base-1.09-pl526_0\n",
      "  perl-xsloader      bioconda/linux-64::perl-xsloader-0.24-pl526_0\n",
      "  sra-tools          bioconda/linux-64::sra-tools-2.11.0-pl5262h314213e_1\n",
      "  tk                 conda-forge/linux-64::tk-8.6.11-h27826a3_1\n",
      "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
      "  zlib               conda-forge/linux-64::zlib-1.2.11-h36c2ea0_1013\n",
      "\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate sra-tools\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda create -y -n sra-tools sra-tools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ee7570f-c35f-462c-9979-daa0999b01f5",
   "metadata": {},
   "source": [
    "## 2.2. Verify `sra-tools` is working\n",
    "\n",
    "Let's verify that the tools found in the package `sra-tools` are working properly. We can do this by running one of the installed commands, `prefetch` like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b2c752f6-108e-4ff9-8538-c49e3177f415",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Usage: prefetch [ options ] [ accessions(s)... ]\n",
      "\n",
      "Parameters:\n",
      "\n",
      "  accessions(s)                    list of accessions to process\n",
      "\n",
      "\n",
      "Options:\n",
      "\n",
      "  -T|--type <file-type>            Specify file type to download. Default: sra\n",
      "  -N|--min-size <size>             Minimum file size to download in KB\n",
      "                                     (inclusive).\n",
      "  -X|--max-size <size>             Maximum file size to download in KB\n",
      "                                     (exclusive). Default: 20G\n",
      "  -f|--force <no|yes|all|ALL>      Force object download - one of: no, yes,\n",
      "                                     all, ALL. no [default]: skip download if\n",
      "                                     the object if found and complete; yes:\n",
      "                                     download it even if it is found and is\n",
      "                                     complete; all: ignore lock files (stale\n",
      "                                     locks or it is being downloaded by\n",
      "                                     another process - use at your own\n",
      "                                     risk!); ALL: ignore lock files, restart\n",
      "                                     download from beginning\n",
      "  -p|--progress                    Show progress\n",
      "  -r|--resume <yes|no>             Resume partial downloads - one of: no, yes\n",
      "                                     [default]\n",
      "  -C|--verify <yes|no>             Verify after download - one of: no, yes\n",
      "                                     [default]\n",
      "  -c|--check-all                   Double-check all refseqs\n",
      "  -o|--output-file <file>          Write file to <file> when downloading\n",
      "                                     single file\n",
      "  -O|--output-directory <directory>\n",
      "                                   Save files to <directory>/\n",
      "     --ngc <path>                  <path> to ngc file\n",
      "     --perm <path>                 <path> to permission file\n",
      "     --location <location>         location in cloud\n",
      "     --cart <path>                 <path> to cart file\n",
      "  -V|--version                     Display the version of the program\n",
      "  -v|--verbose                     Increase the verbosity of the program\n",
      "                                     status messages. Use multiple times for\n",
      "                                     more verbosity.\n",
      "  -L|--log-level <level>           Logging level as number or enum string.\n",
      "                                     One of\n",
      "                                     (fatal|sys|int|err|warn|info|debug) or\n",
      "                                     (0-6) Current/default is warn\n",
      "     --option-file file            Read more options and parameters from the\n",
      "                                     file.\n",
      "  -h|--help                        print this message\n",
      "\n",
      "\"prefetch\" version 2.11.0\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n sra-tools prefetch --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af63b69b-e8bd-4686-b818-9639e6edd259",
   "metadata": {},
   "source": [
    "Notice that I ran by first specifying `conda run -n sra-tools [...]`. The `conda run` part lets you specify a particular environment to activate first before trying to run a tool. It's useful in cases where you are installing each tool in a separate environment (as we are doing here). You can also use `conda activate sra-tools` first and then run `prefetch --help` (but this will not work if running in Jupyter)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8f42647-2ae8-47d5-96b7-67ec2e045a85",
   "metadata": {},
   "source": [
    "## 2.3. Prefetch genomes (`prefetch`)\n",
    "\n",
    "The [prefetch](https://github.com/ncbi/sra-tools/wiki/08.-prefetch-and-fasterq-dump) command that is part of the `sra-tools` package can be used to download and store sequence read data from NCBI's Sequence Read Archive. You can run it like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "222fa969-b9a7-4ef3-a550-088c040ed97c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2022-01-21T19:03:53 prefetch.2.11.0: 1) 'SRR3028792' is found locally\n",
      "2022-01-21T19:03:53 prefetch.2.11.0: 'SRR3028792' has 0 unresolved dependencies\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n sra-tools prefetch SRR3028792"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5efaef5b-0d5a-4bb8-bcb7-cb053bf12bbc",
   "metadata": {},
   "source": [
    "You can even pass multiple accession numbers to `prefetch` like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "870dcff6-61d5-4c6f-b738-86f08e7cbe20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2022-01-21T19:03:55 prefetch.2.11.0: 1) 'SRR3028792' is found locally\n",
      "2022-01-21T19:03:55 prefetch.2.11.0: 'SRR3028792' has 0 unresolved dependencies\n",
      "\n",
      "2022-01-21T19:03:55 prefetch.2.11.0: 2) 'SRR3028793' is found locally\n",
      "2022-01-21T19:03:55 prefetch.2.11.0: 'SRR3028793' has 0 unresolved dependencies\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n sra-tools prefetch SRR3028792 SRR3028793"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7271749-f8d7-4693-bac2-47f1289761f8",
   "metadata": {},
   "source": [
    "The `prefetch` command downloads the data from NCBI and stores on your computer in a special directory by default. You can see which directory using the `srapath` command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "07b56476-0e95-42f8-9d68-7b6d1c159355",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/ncbi/public/sra/SRR3028792.sra\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n sra-tools srapath SRR3028792"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a50478-ec6b-4e49-895b-d6efdbfcc7d2",
   "metadata": {},
   "source": [
    "This tells us that there is a file `SRR3028792.sra` that has been downloaded to the path above which contains the sequence data.\n",
    "\n",
    "For how to know what identifier to use for `prefetch`, this can be found on the NCBI website.\n",
    "\n",
    "For example for `SRR3028792`, you can find this identifier (and the associated biological sample) at https://www.ncbi.nlm.nih.gov/sra/?term=SRR3028792\n",
    "\n",
    "![genome-sra-identifier.png](../images/genome-sra-identifier.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06866e80-a81c-4a6b-bc6b-7d6d50eb5cc7",
   "metadata": {},
   "source": [
    "## 2.4. Convert to fastq (`fasterq-dump`)\n",
    "\n",
    "Once you've prefetched the sequence data you can convert to the **fastq** format using `fasterq-dump` (which is named based on the older `fastq-dump` tool). We need to convert from the `.sra` format to the `.fastq` format as most tools work with **fastq** files (instead of **sra** files, which is specific to NCBI's sequence read archive)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "41443e95-f5d0-4cec-96b9-5759213f3b28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spots read      : 824,262\n",
      "reads read      : 1,648,524\n",
      "reads written   : 1,648,524\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n sra-tools fasterq-dump SRR3028792"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f504e35c-005e-4d6a-81e3-de6ed3a862ee",
   "metadata": {},
   "source": [
    "This will write two `*.fastq` files do your current directory which contain the forward (`_1.fastq`) and reverse (`_2.fastq`) reads.\n",
    "\n",
    "We can use the `ls` command to **l**i**s**t the files and show additional information (such as file size)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fcd6f919-2012-4c4d-9548-c20997eca57a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 396M Jan 21 13:03 SRR3028792_1.fastq\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 398M Jan 21 13:03 SRR3028792_2.fastq\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ls -lh *.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb39e0cb-669b-48bf-a09e-2666c9f75a86",
   "metadata": {},
   "source": [
    "The `396M` part shows us that `SRR3028792_1.fastq` is **396 MB**. The `-lh` part is interpreted as follows: `l` is for **long list**, which means extra information (such as file size) is printed, `h` is for **human readable** which means that file sizes are converted into easier to read units (like MB, GB, etc) instead of being left as bytes. \n",
    "\n",
    "We can compress these files to save on storage space using the `gzip` command.\n",
    "\n",
    "*Tip: To speed this up, you can use `pigz` (parallel gzip) instead of `gzip`, which will take advantage of multiple processing cores on your machine. Just run `pigz *.fastq`. If you don't have `pigz` installed it can be installed with `conda install pigz`.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "27e0db23-e1f0-4c6e-ab2a-a196b4cc2222",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "gzip *.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b2d0bcf6-8489-4cee-94bb-3e2a5207b4a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 138M Jan 21 13:03 \u001b[0m\u001b[01;31mSRR3028792_2.fastq.gz\u001b[0m\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 110M Jan 21 13:03 \u001b[01;31mSRR3028792_1.fastq.gz\u001b[0m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ls -lth *.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d51cbd5-cca1-4658-a997-9c423e35a7fb",
   "metadata": {},
   "source": [
    "Now the files end in `.fastq.gz` instead of `.fastq` and are much smaller (`SRR3028792_1.fastq.gz` is **110 MB** compared to **396 MB** for the uncompressed file)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1becc05-d400-4700-9af9-50d1a75b0f8f",
   "metadata": {},
   "source": [
    "We can also look at the file contents using the `cat` (or `zcat` for files ending in `.gz`) commands along with `head`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2b30427f-5412-4985-b1a0-6b22ff28a0be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR3028792.1 1 length=90\n",
      "CTTCATAATCAGGCGATAAATGCCCACCACTTTAGGCTTTCTGGCGCGGATAGCCTCCCCAATAAAATCTTTACGCGTACGCTTTGCTTC\n",
      "+SRR3028792.1 1 length=90\n",
      "AABC-C--CE,-,C++@,,,<,CCC,BB,:CCE,,,;CCCCC,,6+,+++8,,:,99,,8,,:,,,,<CCEE9C,@+8+8++88C,9CCC\n",
      "@SRR3028792.2 2 length=271\n",
      "ATACACGCCAGCGCGAATATTTATCGCTAGCATCAGGAAAACCAACGAAACAGCCCTCGTTGATTTGCCCCCTGAACAGTGGGCTTTTTTCGCGTACAGATGGCTCCGGCATTAGACCTGTGACATCCCGATACTATTCACATTCCGCAGCTCATACTTCACTTCTGCCCGTAATACGCCTTCGGACCATGCTTACGCATGACGTGTTTATTCATCAGATCATCGTCGATGGGATTCAGCGCCGGGTTAATGCCGCGCGCAATCCACGCCA\n",
      "+SRR3028792.2 2 length=271\n",
      "-86A<F-@F7+C+C@77C,CFF9EE8FE7+C,CE,,,,,,,;@,,8,,,,+,,9:,9,+::,,CEE,CEEDEF7,,:,,,,,,:CEFFEB=+8+++++,,:,,49A?++++8B,,,:??,:,,:,CEFF7+8++:,:A@,>,DDDD,+++88@,@,@CCC,@@B>@,>@D*@**6,5*4*6>?***>:*<,5=>,>*;5=*,,*2*<:C+<CF7A?+++3**;:*9C))7+***29F**9)7<5))8>)*9*8;5;)7)7((-57)5(4:(\n",
      "@SRR3028792.3 3 length=181\n",
      "TGTCGAGATGGTGGCTGAACTGCATCGGGTCTATAGCGCCCAGACTCAGGCGTATGTCCTGTTTAATGAAGAGAGTGCACTTTCGCAAGCGCTTCTGTTGCAACGTCATGGGGAAGAAGATTTTCTTGCCTTTCTCCGGGCGGCGACCGCACCCCTGCGTCACGCACTGTCGACATATCCC\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "\n"
     ]
    }
   ],
   "source": [
    "zcat SRR3028792_1.fastq.gz | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ca4e652-f1c4-4a8c-bbfa-8380a0fa9713",
   "metadata": {},
   "source": [
    "The command `zcat` uncompresses any `*.gz` files and then prints out the contents. Since these are large files we likely only want to look at the first few lines in the file. The command `head` will print out the first 10 lines by default (and ignore all the rest). The `|` character tells bash that we want to take the output of `zcat SRR3028792_1.fastq.gz` (which will print all the lines of a file) and forward it into the `head` command (which will print out the first 10 lines of data and ignore the rest).\n",
    "\n",
    "The end result is we see the first 10 lines of the `SRR3028792_1.fastq.gz` file.\n",
    "\n",
    "We can do the same thing for the reverse reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "69c3e30d-ebae-4711-aad2-33ee7dc1ace7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR3028792.1 1 length=300\n",
      "CTCTCAAACCTTCCTCGTTACTTTTTTCTTTCCGCCTCTCTCCTCCCCCCCCCGCCTCCCCTCTTTTCCCTTTCTCTCCTGCTTCTGCCCCTCTCTCTTCTCCTCCTCTGCCTCTTCCTCCTCCCCCTTACTTCTTCCTCTCTTTCCTCCCCCTCTCCTTCCTCCTCCCCCCCTCTCCCTCCCCCCTTCCCTCCCCTCTCCCTCCCCCCCCTCCCCCCCCCTCCCTCCCTCCTCCCCCCTCTCTCTCCCCTCTCTCCCTCCCCCCTCCCCCCCCTCTCCTCCCCCCCCCCCCTCCCCCCC\n",
      "+SRR3028792.1 1 length=300\n",
      "--,-8,-,;:,;,;,;,:,,,,,6;;@,<@,;;,,8,8;,66;;;,88++++++678,,,,6,,6,,,6,9:?,59,:95,,:9,:,,,,9:+995:9:,:,9,4,999,,9,94,99,,8,,,++84,,,,3,8,,,,,63,,3,,33,+++++,6,,,,,,,,,,,+++*******0******))))*)+*))))))0)))))))))(((((((((((,((,((,(((((((((((((,(())))(,((((.))(((((((((((((((((((,((()((((((((,((((,,(((((\n",
      "@SRR3028792.2 2 length=300\n",
      "TTGCTTCGATTTCTCTCTCCCTTCACCCGGCGCTGACTCCCCTCTCCTCTTCTCTGATGCCTCCTCACTTCATCCTTCCTCCTCCTCCCCCTTCGTATTACTTTCCTCCGTCCCTTCTGCTCTCCTCCCTCTCCCTCTTTTCTCTCTTTCACATTTCTACTTCCGCCCCCCTCTTTCCCCCCACCCACCCCCCTTTTCCGCGTGCCTCTCCACTCTCGCTTTTTCCTTCCTTCTCCTGTTTCTCCCTCTCACTCTTCCCTCTTGCTTTTTTCTTTCTCTTCTACCCCTCTGCCCCTTCCC\n",
      "+SRR3028792.2 2 length=300\n",
      "--,8,;-+-<;,<,C,;,,6,6;,,;6@++8+88++,;C@C,;B,,;,,<;,<;;,,<,,,<,,,:,:C9E,<,:,9,,,<,:,,9:@,,+7,:,:,:9,:,,,9,,,,+9,+,,9,9,,,9499,,,+9,9,,,9,99,49,,,,6,44,7,,,47=,,663=++++6=+66+6=D?+5++++****33903/*333*)))))0)+)1/()/1(/)(((//./.)75)(1.)/7@9))/)11)).((.()).)464).).,)).).),)(,))-.-.65)6)--(,((()).,-((..(\n",
      "@SRR3028792.3 3 length=300\n",
      "CTTCTATTTCTCCCGTTCCTCACTCCGCCTTCCGGTCCCCCCCCGTCTTCACTCAATAACATCTTCTTCCCCCTCTCTTTCCACCCTCACCGCTTCCGCCCCTTCTCTCTCTTCCTTCACCTTCCCCTCCGCCTTCTTCTTCTCCCTCTCTACCCCATCCCTTTCAGCCCCCATCTCTCCTCTCTCTCTTCTCCCCCTCTTCCGCTTCCCCCCACTTCTTCTCTCTCCTTCTCTCTCTTCCCCCTCTCCTTTCTCTCCTCTCTTCCCTCCCCTCCCCTCCCCTCCCTCCTCCTCTCCTCC\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "\n"
     ]
    }
   ],
   "source": [
    "zcat SRR3028792_2.fastq.gz | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec55e0e4-85c5-4d18-82d4-d7af6cb065c8",
   "metadata": {},
   "source": [
    "*Tip: In addition to `head` which prints the first few lines of a file, there is also `tail` which prints the last few lines of a file.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99d8fe83-bf09-41c5-9bc6-0979bf4ce00e",
   "metadata": {},
   "source": [
    "# 3. Quality filter files (`fastp`)\n",
    "\n",
    "The software [fastp](https://github.com/OpenGene/fastp) can be used to generate a report of the quality of the sequence reads as well as remove any poor-quality reads which might cause issues for downstream analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd71b01d-f656-4cd6-91f2-e7dc66c05db6",
   "metadata": {},
   "source": [
    "## 3.1. Install `fastp` using conda\n",
    "\n",
    "We can first install `fastp` using conda."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "faa459cb-e786-4bdf-9f48-d86b734432fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/fastp\n",
      "\n",
      "  added / updated specs:\n",
      "    - fastp\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
      "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
      "  fastp              bioconda/linux-64::fastp-0.23.2-hd36eab0_1\n",
      "  isa-l              conda-forge/linux-64::isa-l-2.30.0-ha770c72_4\n",
      "  libdeflate         conda-forge/linux-64::libdeflate-1.9-h7f98852_0\n",
      "  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11\n",
      "  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11\n",
      "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_11\n",
      "\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate fastp\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda create -y -n fastp fastp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fdce5dd-4fdc-424a-a99b-6c990c730df3",
   "metadata": {},
   "source": [
    "Now let's take a look at all the different options available for running `fastp`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7317080a-1fea-4989-9516-174e9cf0d051",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: fastp [options] ... \n",
      "options:\n",
      "  -i, --in1                            read1 input file name (string [=])\n",
      "  -o, --out1                           read1 output file name (string [=])\n",
      "  -I, --in2                            read2 input file name (string [=])\n",
      "  -O, --out2                           read2 output file name (string [=])\n",
      "      --unpaired1                      for PE input, if read1 passed QC but read2 not, it will be written to unpaired1. Default is to discard it. (string [=])\n",
      "      --unpaired2                      for PE input, if read2 passed QC but read1 not, it will be written to unpaired2. If --unpaired2 is same as --unpaired1 (default mode), both unpaired reads will be written to this same file. (string [=])\n",
      "      --overlapped_out                 for each read pair, output the overlapped region if it has no any mismatched base. (string [=])\n",
      "      --failed_out                     specify the file to store reads that cannot pass the filters. (string [=])\n",
      "  -m, --merge                          for paired-end input, merge each pair of reads into a single read if they are overlapped. The merged reads will be written to the file given by --merged_out, the unmerged reads will be written to the files specified by --out1 and --out2. The merging mode is disabled by default.\n",
      "      --merged_out                     in the merging mode, specify the file name to store merged output, or specify --stdout to stream the merged output (string [=])\n",
      "      --include_unmerged               in the merging mode, write the unmerged or unpaired reads to the file specified by --merge. Disabled by default.\n",
      "  -6, --phred64                        indicate the input is using phred64 scoring (it'll be converted to phred33, so the output will still be phred33)\n",
      "  -z, --compression                    compression level for gzip output (1 ~ 9). 1 is fastest, 9 is smallest, default is 4. (int [=4])\n",
      "      --stdin                          input from STDIN. If the STDIN is interleaved paired-end FASTQ, please also add --interleaved_in.\n",
      "      --stdout                         stream passing-filters reads to STDOUT. This option will result in interleaved FASTQ output for paired-end output. Disabled by default.\n",
      "      --interleaved_in                 indicate that <in1> is an interleaved FASTQ which contains both read1 and read2. Disabled by default.\n",
      "      --reads_to_process               specify how many reads/pairs to be processed. Default 0 means process all reads. (int [=0])\n",
      "      --dont_overwrite                 don't overwrite existing files. Overwritting is allowed by default.\n",
      "      --fix_mgi_id                     the MGI FASTQ ID format is not compatible with many BAM operation tools, enable this option to fix it.\n",
      "  -V, --verbose                        output verbose log information (i.e. when every 1M reads are processed).\n",
      "  -A, --disable_adapter_trimming       adapter trimming is enabled by default. If this option is specified, adapter trimming is disabled\n",
      "  -a, --adapter_sequence               the adapter for read1. For SE data, if not specified, the adapter will be auto-detected. For PE data, this is used if R1/R2 are found not overlapped. (string [=auto])\n",
      "      --adapter_sequence_r2            the adapter for read2 (PE data only). This is used if R1/R2 are found not overlapped. If not specified, it will be the same as <adapter_sequence> (string [=auto])\n",
      "      --adapter_fasta                  specify a FASTA file to trim both read1 and read2 (if PE) by all the sequences in this FASTA file (string [=])\n",
      "      --detect_adapter_for_pe          by default, the auto-detection for adapter is for SE data input only, turn on this option to enable it for PE data.\n",
      "  -f, --trim_front1                    trimming how many bases in front for read1, default is 0 (int [=0])\n",
      "  -t, --trim_tail1                     trimming how many bases in tail for read1, default is 0 (int [=0])\n",
      "  -b, --max_len1                       if read1 is longer than max_len1, then trim read1 at its tail to make it as long as max_len1. Default 0 means no limitation (int [=0])\n",
      "  -F, --trim_front2                    trimming how many bases in front for read2. If it's not specified, it will follow read1's settings (int [=0])\n",
      "  -T, --trim_tail2                     trimming how many bases in tail for read2. If it's not specified, it will follow read1's settings (int [=0])\n",
      "  -B, --max_len2                       if read2 is longer than max_len2, then trim read2 at its tail to make it as long as max_len2. Default 0 means no limitation. If it's not specified, it will follow read1's settings (int [=0])\n",
      "  -D, --dedup                          enable deduplication to drop the duplicated reads/pairs\n",
      "      --dup_calc_accuracy              accuracy level to calculate duplication (1~6), higher level uses more memory (1G, 2G, 4G, 8G, 16G, 24G). Default 1 for no-dedup mode, and 3 for dedup mode. (int [=0])\n",
      "      --dont_eval_duplication          don't evaluate duplication rate to save time and use less memory.\n",
      "  -g, --trim_poly_g                    force polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data\n",
      "      --poly_g_min_len                 the minimum length to detect polyG in the read tail. 10 by default. (int [=10])\n",
      "  -G, --disable_trim_poly_g            disable polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data\n",
      "  -x, --trim_poly_x                    enable polyX trimming in 3' ends.\n",
      "      --poly_x_min_len                 the minimum length to detect polyX in the read tail. 10 by default. (int [=10])\n",
      "  -5, --cut_front                      move a sliding window from front (5') to tail, drop the bases in the window if its mean quality < threshold, stop otherwise.\n",
      "  -3, --cut_tail                       move a sliding window from tail (3') to front, drop the bases in the window if its mean quality < threshold, stop otherwise.\n",
      "  -r, --cut_right                      move a sliding window from front to tail, if meet one window with mean quality < threshold, drop the bases in the window and the right part, and then stop.\n",
      "  -W, --cut_window_size                the window size option shared by cut_front, cut_tail or cut_sliding. Range: 1~1000, default: 4 (int [=4])\n",
      "  -M, --cut_mean_quality               the mean quality requirement option shared by cut_front, cut_tail or cut_sliding. Range: 1~36 default: 20 (Q20) (int [=20])\n",
      "      --cut_front_window_size          the window size option of cut_front, default to cut_window_size if not specified (int [=4])\n",
      "      --cut_front_mean_quality         the mean quality requirement option for cut_front, default to cut_mean_quality if not specified (int [=20])\n",
      "      --cut_tail_window_size           the window size option of cut_tail, default to cut_window_size if not specified (int [=4])\n",
      "      --cut_tail_mean_quality          the mean quality requirement option for cut_tail, default to cut_mean_quality if not specified (int [=20])\n",
      "      --cut_right_window_size          the window size option of cut_right, default to cut_window_size if not specified (int [=4])\n",
      "      --cut_right_mean_quality         the mean quality requirement option for cut_right, default to cut_mean_quality if not specified (int [=20])\n",
      "  -Q, --disable_quality_filtering      quality filtering is enabled by default. If this option is specified, quality filtering is disabled\n",
      "  -q, --qualified_quality_phred        the quality value that a base is qualified. Default 15 means phred quality >=Q15 is qualified. (int [=15])\n",
      "  -u, --unqualified_percent_limit      how many percents of bases are allowed to be unqualified (0~100). Default 40 means 40% (int [=40])\n",
      "  -n, --n_base_limit                   if one read's number of N base is >n_base_limit, then this read/pair is discarded. Default is 5 (int [=5])\n",
      "  -e, --average_qual                   if one read's average quality score <avg_qual, then this read/pair is discarded. Default 0 means no requirement (int [=0])\n",
      "  -L, --disable_length_filtering       length filtering is enabled by default. If this option is specified, length filtering is disabled\n",
      "  -l, --length_required                reads shorter than length_required will be discarded, default is 15. (int [=15])\n",
      "      --length_limit                   reads longer than length_limit will be discarded, default 0 means no limitation. (int [=0])\n",
      "  -y, --low_complexity_filter          enable low complexity filter. The complexity is defined as the percentage of base that is different from its next base (base[i] != base[i+1]).\n",
      "  -Y, --complexity_threshold           the threshold for low complexity filter (0~100). Default is 30, which means 30% complexity is required. (int [=30])\n",
      "      --filter_by_index1               specify a file contains a list of barcodes of index1 to be filtered out, one barcode per line (string [=])\n",
      "      --filter_by_index2               specify a file contains a list of barcodes of index2 to be filtered out, one barcode per line (string [=])\n",
      "      --filter_by_index_threshold      the allowed difference of index barcode for index filtering, default 0 means completely identical. (int [=0])\n",
      "  -c, --correction                     enable base correction in overlapped regions (only for PE data), default is disabled\n",
      "      --overlap_len_require            the minimum length to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. 30 by default. (int [=30])\n",
      "      --overlap_diff_limit             the maximum number of mismatched bases to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. 5 by default. (int [=5])\n",
      "      --overlap_diff_percent_limit     the maximum percentage of mismatched bases to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. Default 20 means 20%. (int [=20])\n",
      "  -U, --umi                            enable unique molecular identifier (UMI) preprocessing\n",
      "      --umi_loc                        specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none (string [=])\n",
      "      --umi_len                        if the UMI is in read1/read2, its length should be provided (int [=0])\n",
      "      --umi_prefix                     if specified, an underline will be used to connect prefix and UMI (i.e. prefix=UMI, UMI=AATTCG, final=UMI_AATTCG). No prefix by default (string [=])\n",
      "      --umi_skip                       if the UMI is in read1/read2, fastp can skip several bases following UMI, default is 0 (int [=0])\n",
      "  -p, --overrepresentation_analysis    enable overrepresented sequence analysis.\n",
      "  -P, --overrepresentation_sampling    one in (--overrepresentation_sampling) reads will be computed for overrepresentation analysis (1~10000), smaller is slower, default is 20. (int [=20])\n",
      "  -j, --json                           the json format report file name (string [=fastp.json])\n",
      "  -h, --html                           the html format report file name (string [=fastp.html])\n",
      "  -R, --report_title                   should be quoted with ' or \", default is \"fastp report\" (string [=fastp report])\n",
      "  -w, --thread                         worker thread number, default is 3 (int [=3])\n",
      "  -s, --split                          split output by limiting total split file number with this option (2~999), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (int [=0])\n",
      "  -S, --split_by_lines                 split output by limiting lines of each file with this option(>=1000), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (long [=0])\n",
      "  -d, --split_prefix_digits            the digits for the sequential number padding (1~10), default is 4, so the filename will be padded as 0001.xxx, 0 to disable padding (int [=4])\n",
      "      --cut_by_quality5                DEPRECATED, use --cut_front instead.\n",
      "      --cut_by_quality3                DEPRECATED, use --cut_tail instead.\n",
      "      --cut_by_quality_aggressive      DEPRECATED, use --cut_right instead.\n",
      "      --discard_unmerged               DEPRECATED, no effect now, see the introduction for merging.\n",
      "  -?, --help                           print this message\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n fastp fastp --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e5630fe-c25a-491d-b1c0-888866bcb880",
   "metadata": {},
   "source": [
    "## 3.2. Evaluate quality of sequence reads\n",
    "\n",
    "To run `fastp` on our set of sequence reads, we can pass in the `*.fastq.gz` files to the command and specify the output file names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b7e7f612-fca1-4242-80b8-428e1940a64b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read1 before filtering:\n",
      "total reads: 824262\n",
      "total bases: 176567188\n",
      "Q20 bases: 171950867(97.3855%)\n",
      "Q30 bases: 163432314(92.561%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 824262\n",
      "total bases: 177556496\n",
      "Q20 bases: 153139830(86.2485%)\n",
      "Q30 bases: 134061609(75.5036%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 788353\n",
      "total bases: 166850080\n",
      "Q20 bases: 163779407(98.1596%)\n",
      "Q30 bases: 156905038(94.0395%)\n",
      "\n",
      "Read2 after filtering:\n",
      "total reads: 788353\n",
      "total bases: 166698912\n",
      "Q20 bases: 149106629(89.4467%)\n",
      "Q30 bases: 132133429(79.2647%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 1576706\n",
      "reads failed due to low quality: 71818\n",
      "reads failed due to too many N: 0\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 46646\n",
      "bases trimmed due to adapters: 1033979\n",
      "\n",
      "Duplication rate: 0.665322%\n",
      "\n",
      "Insert size peak (evaluated by paired-end reads): 35\n",
      "\n",
      "JSON report: SRR3028792.json\n",
      "HTML report: SRR3028792.html\n",
      "\n",
      "fastp --in1 SRR3028792_1.fastq.gz --in2 SRR3028792_2.fastq.gz --out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fastq.gz --html SRR3028792.html --json SRR3028792.json \n",
      "fastp v0.23.2, time used: 7 seconds\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n fastp fastp --in1 SRR3028792_1.fastq.gz --in2 SRR3028792_2.fastq.gz --out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fastq.gz --html SRR3028792.html --json SRR3028792.json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa885a3d-82e4-4fdb-a049-40661901429f",
   "metadata": {},
   "source": [
    "The meaning of the following command-line arguments is as follows:\n",
    "\n",
    "* `--in1 SRR3028792_1.fastq.gz --in2 SRR3028792_2.fastq.gz`: The first pair of fastq reads (`--in1`) and second pair (`--in2`) to examine.\n",
    "* `--out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fastq.gz`: The output file names for reads that pass the default quality filters, `--out1` for the first pair and `--out2` for the second pair.\n",
    "* `--html SRR3028792.html`: The name of the HTML report, which can be opened in a web browser.\n",
    "* `--json SRR3028792.json`: The name of the JSON report (an alternative format to HTML).\n",
    "\n",
    "Once `fastp` is finished you can open up the `SRR3028792.html` report in any web browser. In Jupyter you can right-click and select **Open in New Browser Tab**:\n",
    "\n",
    "![jupyter-fastp-report.png](../images/jupyter-fastp-report.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "000237d7-e164-41d4-87ec-fb6cfe8033c4",
   "metadata": {},
   "source": [
    "# 4. Genome assembly (`skesa`)\n",
    "\n",
    "Now that we have removed poor-quality reads from our **fastq** files we can move on to genome assembly, which attempts to fit all the smaller sequence reads in the **fastq** files together to produce longer contiguous sequences (contigs). Ideally you would have one contig per chromosome/plasmid but often chromosomes/plasmids will be broken up into smaller contigs, normally at repetitive regions on the genome that are longer than the read length.\n",
    "\n",
    "## 4.1. Install skesa\n",
    "\n",
    "We will use [skesa](https://github.com/ncbi/SKESA) to assemble the sequence reads. Skesa is a *de-novo* genome assembler that is designed specifically for microbial genomes. The first step is installing the software using conda."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "da879b03-092a-46d6-92eb-3ec9870f7b8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/skesa\n",
      "\n",
      "  added / updated specs:\n",
      "    - skesa\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
      "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
      "  boost              conda-forge/linux-64::boost-1.70.0-py38h9de70de_1\n",
      "  boost-cpp          conda-forge/linux-64::boost-cpp-1.70.0-h7b93d67_3\n",
      "  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4\n",
      "  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0\n",
      "  icu                conda-forge/linux-64::icu-67.1-he1b5a44_0\n",
      "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.36.1-hea4e1c9_2\n",
      "  libblas            conda-forge/linux-64::libblas-3.9.0-13_linux64_openblas\n",
      "  libcblas           conda-forge/linux-64::libcblas-3.9.0-13_linux64_openblas\n",
      "  libffi             conda-forge/linux-64::libffi-3.4.2-h7f98852_5\n",
      "  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11\n",
      "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-11.2.0-h69a702a_11\n",
      "  libgfortran5       conda-forge/linux-64::libgfortran5-11.2.0-h5c6108e_11\n",
      "  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11\n",
      "  liblapack          conda-forge/linux-64::liblapack-3.9.0-13_linux64_openblas\n",
      "  libnsl             conda-forge/linux-64::libnsl-2.0.0-h7f98852_0\n",
      "  libopenblas        conda-forge/linux-64::libopenblas-0.3.18-pthreads_h8fe5266_0\n",
      "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_11\n",
      "  libzlib            conda-forge/linux-64::libzlib-1.2.11-h36c2ea0_1013\n",
      "  lz4-c              conda-forge/linux-64::lz4-c-1.9.3-h9c3ff4c_1\n",
      "  ncurses            conda-forge/linux-64::ncurses-6.3-h9c3ff4c_0\n",
      "  numpy              conda-forge/linux-64::numpy-1.22.0-py38h6ae9a64_1\n",
      "  openssl            conda-forge/linux-64::openssl-3.0.0-h7f98852_2\n",
      "  pip                conda-forge/noarch::pip-21.3.1-pyhd8ed1ab_0\n",
      "  python             conda-forge/linux-64::python-3.8.12-hf930737_2_cpython\n",
      "  python_abi         conda-forge/linux-64::python_abi-3.8-2_cp38\n",
      "  readline           conda-forge/linux-64::readline-8.1-h46c0cb4_0\n",
      "  setuptools         conda-forge/linux-64::setuptools-60.5.0-py38h578d9bd_0\n",
      "  skesa              bioconda/linux-64::skesa-2.4.0-he1c1bb9_0\n",
      "  sqlite             conda-forge/linux-64::sqlite-3.37.0-h9cd32fc_0\n",
      "  tk                 conda-forge/linux-64::tk-8.6.11-h27826a3_1\n",
      "  wheel              conda-forge/noarch::wheel-0.37.1-pyhd8ed1ab_0\n",
      "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
      "  zlib               conda-forge/linux-64::zlib-1.2.11-h36c2ea0_1013\n",
      "  zstd               conda-forge/linux-64::zstd-1.4.9-ha95c52a_0\n",
      "\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate skesa\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda create -y -n skesa skesa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4fe4cc6e-375a-426f-b070-d66c6b491057",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "General options:\n",
      "  -h [ --help ]                 Produce help message\n",
      "  -v [ --version ]              Print version\n",
      "  --cores arg (=0)              Number of cores to use (default all) [integer]\n",
      "  --memory arg (=32)            Memory available (GB, only for sorted counter) \n",
      "                                [integer]\n",
      "  --hash_count                  Use hash counter [flag]\n",
      "  --estimated_kmers arg (=100)  Estimated number of unique kmers for bloom \n",
      "                                filter (millions, only for hash counter) \n",
      "                                [integer]\n",
      "  --skip_bloom_filter           Don't do bloom filter; use --estimated_kmers as\n",
      "                                the hash table size (only for hash counter) \n",
      "                                [flag]\n",
      "\n",
      "Input/output options : at least one input providing reads for assembly must be specified:\n",
      "  --reads arg                   Input fasta/fastq file(s) for reads (could be \n",
      "                                used multiple times for different runs, could \n",
      "                                be gzipped) [string]\n",
      "  --use_paired_ends             Indicates that a single (not comma separated) \n",
      "                                fasta/fastq file contains paired reads [flag]\n",
      "  --contigs_out arg             Output file for contigs (stdout if not \n",
      "                                specified) [string]\n",
      "\n",
      "Assembly options:\n",
      "  --kmer arg (=21)              Minimal kmer length for assembly [integer]\n",
      "  --min_count arg               Minimal count for kmers retained for comparing \n",
      "                                alternate choices [integer]\n",
      "  --max_kmer arg                Maximal kmer length for assembly [integer]\n",
      "  --max_kmer_count arg          Minimum acceptable average count for estimating\n",
      "                                the maximal kmer length in reads [integer]\n",
      "  --vector_percent arg (=0.05)  Count for  vectors as a fraction of the read \n",
      "                                number (1. disables) [float (0,1]]\n",
      "  --insert_size arg             Expected insert size for paired reads (if not \n",
      "                                provided, it will be estimated) [integer]\n",
      "  --steps arg (=11)             Number of assembly iterations from minimal to \n",
      "                                maximal kmer length in reads [integer]\n",
      "  --fraction arg (=0.1)         Maximum noise to signal ratio acceptable for \n",
      "                                extension [float [0,1)]\n",
      "  --max_snp_len arg (=150)      Maximal snp length [integer]\n",
      "  --min_contig arg (=200)       Minimal contig length reported in output \n",
      "                                [integer]\n",
      "  --allow_snps                  Allow additional step for snp discovery [flag]\n",
      "\n",
      "Debugging options:\n",
      "  --force_single_ends           Don't use paired-end information [flag]\n",
      "  --seeds arg                   Input file with seeds [string]\n",
      "  --all arg                     Output fasta for each iteration [string]\n",
      "  --dbg_out arg                 Output kmer file [string]\n",
      "  --hist arg                    File for histogram [string]\n",
      "  --connected_reads arg         File for connected paired reads [string]\n",
      "\n",
      "\n",
      "skesa --help \n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n skesa skesa --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63b3e0c9-5349-48a6-b6d0-42219db8cc82",
   "metadata": {},
   "source": [
    "## 4.2. Assemble genome\n",
    "\n",
    "Next we will run `skesa` on our filtered **fastq** files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "783bc907-e4e4-40cc-8190-fbe03af9d060",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skesa --reads SRR3028792_1.fp.fastq.gz,SRR3028792_2.fp.fastq.gz --contigs_out SRR3028792.fasta \n",
      "\n",
      "Total mates: 1576706 Paired reads: 788353\n",
      "Reads acquired in  14.197494s wall, 13.940000s user + 0.230000s system = 14.170000s CPU (99.8%)\n",
      "\n",
      "Kmer len: 19\n",
      "Raw kmers: 305168274 Memory needed (GB): 5.85923 Memory available (GB): 29.8726 1 cycle(s) will be performed\n",
      "Distinct kmers: 172827\n",
      "Kmer count in  3.678642s wall, 148.840000s user + 9.670000s system = 158.510000s CPU (4308.9%)\n",
      "Uniq kmers merging in  0.172947s wall, 0.060000s user + 0.480000s system = 0.540000s CPU (312.2%)\n",
      "Adapters: 0 Reads before: 1576706 Sequence before: 333548982 Reads after: 1576706 Sequence after: 333548982 Reads clipped: 0\n",
      "Adapters clipped in  3.858329s wall, 148.900000s user + 10.150000s system = 159.050000s CPU (4122.3%)\n",
      "\n",
      "Kmer len: 21\n",
      "Raw kmers: 302014862 Memory needed (GB): 5.79869 Memory available (GB): 29.783 1 cycle(s) will be performed\n",
      "Distinct kmers: 5886038\n",
      "Kmer count in  3.480091s wall, 151.330000s user + 9.160000s system = 160.490000s CPU (4611.7%)\n",
      "Uniq kmers merging in  0.539677s wall, 1.010000s user + 1.610000s system = 2.620000s CPU (485.5%)\n",
      "Kmers branching in  0.523497s wall, 20.020000s user + 0.020000s system = 20.040000s CPU (3828.1%)\n",
      "\n",
      "Average read length: 212\n",
      "Genome size estimate: 4388683\n",
      "\n",
      "Kmer: 21 Graph size: 5886038 Contigs in: 0\n",
      "Valley: 7\n",
      "\n",
      "Mark used kmers in  0.000003s wall, 0.000000s user + 0.000000s system = 0.000000s CPU (n/a%)\n",
      "Kmers in multiple/single contigs: 0 0\n",
      "Fragments before: 1868 4803175\n",
      "Fragments after: 1645 4803175\n",
      "New seeds: 1344\n",
      "New seeds in  1.179451s wall, 57.720000s user + 0.510000s system = 58.230000s CPU (4937.0%)\n",
      "Fragments before: 2688 56448\n",
      "Fragments after: 2688 56448\n",
      "Connectors: 0 Extenders: 2688\n",
      "Connections and extensions in  0.116862s wall, 4.470000s user + 0.140000s system = 4.610000s CPU (3944.8%)\n",
      "Contigs out: 1344 Genome: 4756810 N50: 7553 L50: 187\n",
      "Assembled in  1.319606s wall, 62.210000s user + 0.650000s system = 62.860000s CPU (4763.5%)\n",
      "\n",
      "\n",
      "Kmer len: 173\n",
      "Raw kmers: 99928415 Memory needed (GB): 6.71519 Memory available (GB): 29.678 1 cycle(s) will be performed\n",
      "Distinct kmers: 5408481\n",
      "Kmer count in  3.255032s wall, 140.660000s user + 8.410000s system = 149.070000s CPU (4579.7%)\n",
      "Uniq kmers merging in  1.063943s wall, 1.750000s user + 4.120000s system = 5.870000s CPU (551.7%)\n",
      "\n",
      "Average count: 50.4151 Max kmer: 173\n",
      "Connected: 9206 ambiguously connected: 286 from 10000 mate pairs\n",
      "Connect pairs in  1.625097s wall, 72.390000s user + 0.060000s system = 72.450000s CPU (4458.2%)\n",
      "\n",
      "N50 for inserts: 363\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 311364\n",
      "Reads for connection: 158848\n",
      "Internal reads: 23033\n",
      "Reads cleaned in  0.593117s wall, 21.340000s user + 0.460000s system = 21.800000s CPU (3675.5%)\n",
      "\n",
      "Kmer len: 35\n",
      "Raw kmers: 59168172 Memory needed (GB): 1.70404 Memory available (GB): 29.7137 1 cycle(s) will be performed\n",
      "Distinct kmers: 1815510\n",
      "Kmer count in  0.842707s wall, 30.680000s user + 1.810000s system = 32.490000s CPU (3855.4%)\n",
      "Uniq kmers merging in  0.146724s wall, 0.390000s user + 0.120000s system = 0.510000s CPU (347.6%)\n",
      "Kmers branching in  0.171338s wall, 6.850000s user + 0.010000s system = 6.860000s CPU (4003.8%)\n",
      "Kmer: 35 Graph size: 1815510 Contigs in: 1344\n",
      "Valley: 18\n",
      "\n",
      "Mark used kmers in  0.089449s wall, 4.180000s user + 0.020000s system = 4.200000s CPU (4695.4%)\n",
      "Kmers in multiple/single contigs: 0 1419736\n",
      "Fragments before: 1775 106879\n",
      "Fragments after: 1739 106879\n",
      "New seeds: 62\n",
      "New seeds in  0.330815s wall, 13.680000s user + 0.000000s system = 13.680000s CPU (4135.2%)\n",
      "Fragments before: 2812 93582\n",
      "Fragments after: 1747 93582\n",
      "Connectors: 1065 Extenders: 682\n",
      "Connections and extensions in  0.174319s wall, 5.250000s user + 0.010000s system = 5.260000s CPU (3017.4%)\n",
      "Contigs out: 342 Genome: 4801583 N50: 39935 L50: 37\n",
      "Assembled in  0.614917s wall, 23.130000s user + 0.030000s system = 23.160000s CPU (3766.4%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 106218\n",
      "Reads for connection: 63592\n",
      "Internal reads: 6675\n",
      "Reads cleaned in  0.548484s wall, 15.440000s user + 0.140000s system = 15.580000s CPU (2840.6%)\n",
      "\n",
      "Kmer len: 51\n",
      "Raw kmers: 17693418 Memory needed (GB): 0.50957 Memory available (GB): 29.8143 1 cycle(s) will be performed\n",
      "Distinct kmers: 514372\n",
      "Kmer count in  0.225790s wall, 8.150000s user + 0.170000s system = 8.320000s CPU (3684.8%)\n",
      "Uniq kmers merging in  0.053534s wall, 0.150000s user + 0.030000s system = 0.180000s CPU (336.2%)\n",
      "Kmers branching in  0.052253s wall, 2.100000s user + 0.010000s system = 2.110000s CPU (4038.1%)\n",
      "Kmer: 51 Graph size: 514372 Contigs in: 342\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.086066s wall, 2.930000s user + 0.000000s system = 2.930000s CPU (3404.4%)\n",
      "Kmers in multiple/single contigs: 0 340226\n",
      "Fragments before: 762 53318\n",
      "Fragments after: 700 53318\n",
      "New seeds: 19\n",
      "New seeds in  0.181229s wall, 7.970000s user + 0.010000s system = 7.980000s CPU (4403.3%)\n",
      "Fragments before: 720 35276\n",
      "Fragments after: 581 35276\n",
      "Connectors: 139 Extenders: 442\n",
      "Connections and extensions in  0.222347s wall, 5.140000s user + 0.030000s system = 5.170000s CPU (2325.2%)\n",
      "Contigs out: 222 Genome: 4814548 N50: 66983 L50: 25\n",
      "Assembled in  0.496901s wall, 16.050000s user + 0.040000s system = 16.090000s CPU (3238.1%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 70428\n",
      "Reads for connection: 44548\n",
      "Internal reads: 4457\n",
      "Reads cleaned in  0.535426s wall, 10.870000s user + 0.150000s system = 11.020000s CPU (2058.2%)\n",
      "\n",
      "Kmer len: 67\n",
      "Raw kmers: 10963309 Memory needed (GB): 0.420991 Memory available (GB): 29.8166 1 cycle(s) will be performed\n",
      "Distinct kmers: 329566\n",
      "Kmer count in  0.182317s wall, 6.420000s user + 0.040000s system = 6.460000s CPU (3543.3%)\n",
      "Uniq kmers merging in  0.053211s wall, 0.080000s user + 0.070000s system = 0.150000s CPU (281.9%)\n",
      "Kmers branching in  0.041127s wall, 1.730000s user + 0.020000s system = 1.750000s CPU (4255.1%)\n",
      "Kmer: 67 Graph size: 329566 Contigs in: 222\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.092026s wall, 2.860000s user + 0.000000s system = 2.860000s CPU (3107.8%)\n",
      "Kmers in multiple/single contigs: 0 201838\n",
      "Fragments before: 516 39362\n",
      "Fragments after: 481 39362\n",
      "New seeds: 5\n",
      "New seeds in  0.128879s wall, 5.460000s user + 0.010000s system = 5.470000s CPU (4244.3%)\n",
      "Fragments before: 452 27425\n",
      "Fragments after: 410 27425\n",
      "Connectors: 42 Extenders: 368\n",
      "Connections and extensions in  0.207801s wall, 5.230000s user + 0.040000s system = 5.270000s CPU (2536.1%)\n",
      "Contigs out: 185 Genome: 4820584 N50: 78027 L50: 21\n",
      "Assembled in  0.433071s wall, 13.560000s user + 0.050000s system = 13.610000s CPU (3142.7%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 57164\n",
      "Reads for connection: 37310\n",
      "Internal reads: 3939\n",
      "Reads cleaned in  0.543562s wall, 10.680000s user + 0.160000s system = 10.840000s CPU (1994.3%)\n",
      "\n",
      "Kmer len: 81\n",
      "Raw kmers: 8614301 Memory needed (GB): 0.330789 Memory available (GB): 29.8041 1 cycle(s) will be performed\n",
      "Distinct kmers: 270496\n",
      "Kmer count in  0.135363s wall, 4.980000s user + 0.060000s system = 5.040000s CPU (3723.3%)\n",
      "Uniq kmers merging in  0.046974s wall, 0.070000s user + 0.050000s system = 0.120000s CPU (255.5%)\n",
      "Kmers branching in  0.037875s wall, 1.520000s user + 0.010000s system = 1.530000s CPU (4039.6%)\n",
      "Kmer: 81 Graph size: 270496 Contigs in: 185\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.095485s wall, 2.830000s user + 0.000000s system = 2.830000s CPU (2963.8%)\n",
      "Kmers in multiple/single contigs: 0 160387\n",
      "Fragments before: 414 32188\n",
      "Fragments after: 372 32188\n",
      "New seeds: 2\n",
      "New seeds in  0.101329s wall, 4.200000s user + 0.010000s system = 4.210000s CPU (4154.8%)\n",
      "Fragments before: 358 25436\n",
      "Fragments after: 327 25436\n",
      "Connectors: 31 Extenders: 296\n",
      "Connections and extensions in  0.285320s wall, 5.040000s user + 0.020000s system = 5.060000s CPU (1773.4%)\n",
      "Contigs out: 149 Genome: 4824918 N50: 100845 L50: 17\n",
      "Assembled in  0.485746s wall, 12.070000s user + 0.030000s system = 12.100000s CPU (2491.0%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 48264\n",
      "Reads for connection: 33700\n",
      "Internal reads: 3321\n",
      "Reads cleaned in  0.592399s wall, 10.060000s user + 0.160000s system = 10.220000s CPU (1725.2%)\n",
      "\n",
      "Kmer len: 97\n",
      "Raw kmers: 6753463 Memory needed (GB): 0.324166 Memory available (GB): 29.7915 1 cycle(s) will be performed\n",
      "Distinct kmers: 222243\n",
      "Kmer count in  0.151276s wall, 5.280000s user + 0.090000s system = 5.370000s CPU (3549.8%)\n",
      "Uniq kmers merging in  0.044680s wall, 0.090000s user + 0.040000s system = 0.130000s CPU (291.0%)\n",
      "Kmers branching in  0.036702s wall, 1.390000s user + 0.000000s system = 1.390000s CPU (3787.3%)\n",
      "Kmer: 97 Graph size: 222243 Contigs in: 149\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.125798s wall, 2.960000s user + 0.010000s system = 2.970000s CPU (2360.9%)\n",
      "Kmers in multiple/single contigs: 0 127122\n",
      "Fragments before: 316 28341\n",
      "Fragments after: 285 28341\n",
      "New seeds: 2\n",
      "New seeds in  0.089157s wall, 3.840000s user + 0.010000s system = 3.850000s CPU (4318.2%)\n",
      "Fragments before: 284 23931\n",
      "Fragments after: 260 23931\n",
      "Connectors: 24 Extenders: 236\n",
      "Connections and extensions in  0.436030s wall, 4.870000s user + 0.030000s system = 4.900000s CPU (1123.8%)\n",
      "Contigs out: 119 Genome: 4829069 N50: 132289 L50: 13\n",
      "Assembled in  0.654587s wall, 11.670000s user + 0.050000s system = 11.720000s CPU (1790.4%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 41477\n",
      "Reads for connection: 31082\n",
      "Internal reads: 2868\n",
      "Reads cleaned in  0.718913s wall, 10.050000s user + 0.140000s system = 10.190000s CPU (1417.4%)\n",
      "\n",
      "Kmer len: 111\n",
      "Raw kmers: 5486019 Memory needed (GB): 0.263329 Memory available (GB): 29.7782 1 cycle(s) will be performed\n",
      "Distinct kmers: 184837\n",
      "Kmer count in  0.155154s wall, 4.440000s user + 0.100000s system = 4.540000s CPU (2926.1%)\n",
      "Uniq kmers merging in  0.046972s wall, 0.060000s user + 0.060000s system = 0.120000s CPU (255.5%)\n",
      "Kmers branching in  0.030197s wall, 1.180000s user + 0.000000s system = 1.180000s CPU (3907.7%)\n",
      "Kmer: 111 Graph size: 184837 Contigs in: 119\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.155756s wall, 2.560000s user + 0.000000s system = 2.560000s CPU (1643.6%)\n",
      "Kmers in multiple/single contigs: 0 101546\n",
      "Fragments before: 268 24358\n",
      "Fragments after: 241 24358\n",
      "New seeds: 1\n",
      "New seeds in  0.077967s wall, 3.130000s user + 0.010000s system = 3.140000s CPU (4027.3%)\n",
      "Fragments before: 233 20168\n",
      "Fragments after: 216 20168\n",
      "Connectors: 17 Extenders: 199\n",
      "Connections and extensions in  0.461484s wall, 4.480000s user + 0.060000s system = 4.540000s CPU (983.8%)\n",
      "Contigs out: 101 Genome: 4832150 N50: 149715 L50: 12\n",
      "Assembled in  0.698192s wall, 10.180000s user + 0.070000s system = 10.250000s CPU (1468.1%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 36740\n",
      "Reads for connection: 29274\n",
      "Internal reads: 2514\n",
      "Reads cleaned in  0.800916s wall, 9.320000s user + 0.270000s system = 9.590000s CPU (1197.4%)\n",
      "\n",
      "Kmer len: 127\n",
      "Raw kmers: 4492990 Memory needed (GB): 0.215664 Memory available (GB): 29.7664 1 cycle(s) will be performed\n",
      "Distinct kmers: 159227\n",
      "Kmer count in  0.145391s wall, 3.740000s user + 0.050000s system = 3.790000s CPU (2606.8%)\n",
      "Uniq kmers merging in  0.046569s wall, 0.060000s user + 0.060000s system = 0.120000s CPU (257.7%)\n",
      "Kmers branching in  0.027164s wall, 1.080000s user + 0.010000s system = 1.090000s CPU (4012.6%)\n",
      "Kmer: 127 Graph size: 159227 Contigs in: 101\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.173067s wall, 2.460000s user + 0.010000s system = 2.470000s CPU (1427.2%)\n",
      "Kmers in multiple/single contigs: 0 84892\n",
      "Fragments before: 235 21802\n",
      "Fragments after: 201 21802\n",
      "New seeds: 0\n",
      "New seeds in  0.070314s wall, 2.920000s user + 0.000000s system = 2.920000s CPU (4152.8%)\n",
      "Fragments before: 191 19078\n",
      "Fragments after: 180 19078\n",
      "Connectors: 11 Extenders: 169\n",
      "Connections and extensions in  0.444060s wall, 4.530000s user + 0.030000s system = 4.560000s CPU (1026.9%)\n",
      "Contigs out: 87 Genome: 4834786 N50: 160303 L50: 10\n",
      "Assembled in  0.690054s wall, 9.920000s user + 0.040000s system = 9.960000s CPU (1443.4%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 32363\n",
      "Reads for connection: 27660\n",
      "Internal reads: 2233\n",
      "Reads cleaned in  0.995454s wall, 8.200000s user + 0.290000s system = 8.490000s CPU (852.9%)\n",
      "\n",
      "Kmer len: 143\n",
      "Raw kmers: 3670932 Memory needed (GB): 0.211446 Memory available (GB): 29.7554 1 cycle(s) will be performed\n",
      "Distinct kmers: 138396\n",
      "Kmer count in  0.130033s wall, 3.360000s user + 0.070000s system = 3.430000s CPU (2637.8%)\n",
      "Uniq kmers merging in  0.059585s wall, 0.070000s user + 0.060000s system = 0.130000s CPU (218.2%)\n",
      "Kmers branching in  0.027481s wall, 1.050000s user + 0.000000s system = 1.050000s CPU (3820.9%)\n",
      "Kmer: 143 Graph size: 138396 Contigs in: 87\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.239295s wall, 2.500000s user + 0.030000s system = 2.530000s CPU (1057.3%)\n",
      "Kmers in multiple/single contigs: 0 72856\n",
      "Fragments before: 209 20112\n",
      "Fragments after: 176 20112\n",
      "New seeds: 0\n",
      "New seeds in  0.062507s wall, 2.330000s user + 0.000000s system = 2.330000s CPU (3727.6%)\n",
      "Fragments before: 162 16610\n",
      "Fragments after: 153 16610\n",
      "Connectors: 9 Extenders: 144\n",
      "Connections and extensions in  0.491708s wall, 4.420000s user + 0.050000s system = 4.470000s CPU (909.1%)\n",
      "Contigs out: 75 Genome: 4836364 N50: 161583 L50: 9\n",
      "Assembled in  0.795604s wall, 9.260000s user + 0.080000s system = 9.340000s CPU (1174.0%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 27994\n",
      "Reads for connection: 26352\n",
      "Internal reads: 1868\n",
      "Reads cleaned in  1.007378s wall, 8.780000s user + 0.320000s system = 9.100000s CPU (903.3%)\n",
      "\n",
      "Kmer len: 157\n",
      "Raw kmers: 2971015 Memory needed (GB): 0.17113 Memory available (GB): 29.7441 1 cycle(s) will be performed\n",
      "Distinct kmers: 118719\n",
      "Kmer count in  0.145876s wall, 2.690000s user + 0.140000s system = 2.830000s CPU (1940.0%)\n",
      "Uniq kmers merging in  0.042768s wall, 0.040000s user + 0.060000s system = 0.100000s CPU (233.8%)\n",
      "Kmers branching in  0.026669s wall, 0.910000s user + 0.010000s system = 0.920000s CPU (3449.6%)\n",
      "Kmer: 157 Graph size: 118719 Contigs in: 75\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.205247s wall, 2.430000s user + 0.000000s system = 2.430000s CPU (1183.9%)\n",
      "Kmers in multiple/single contigs: 0 60930\n",
      "Fragments before: 202 19385\n",
      "Fragments after: 163 19385\n",
      "New seeds: 0\n",
      "New seeds in  0.059320s wall, 2.110000s user + 0.010000s system = 2.120000s CPU (3573.9%)\n",
      "Fragments before: 141 14886\n",
      "Fragments after: 137 14886\n",
      "Connectors: 4 Extenders: 133\n",
      "Connections and extensions in  0.478595s wall, 4.260000s user + 0.080000s system = 4.340000s CPU (906.8%)\n",
      "Contigs out: 70 Genome: 4837350 N50: 197750 L50: 9\n",
      "Assembled in  0.744972s wall, 8.800000s user + 0.090000s system = 8.890000s CPU (1193.3%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 25999\n",
      "Reads for connection: 25942\n",
      "Internal reads: 1732\n",
      "Reads cleaned in  1.016273s wall, 7.890000s user + 0.300000s system = 8.190000s CPU (805.9%)\n",
      "\n",
      "Kmer len: 173\n",
      "Raw kmers: 2489781 Memory needed (GB): 0.167313 Memory available (GB): 29.7338 1 cycle(s) will be performed\n",
      "Distinct kmers: 107041\n",
      "Kmer count in  0.144334s wall, 2.630000s user + 0.150000s system = 2.780000s CPU (1926.1%)\n",
      "Uniq kmers merging in  0.037959s wall, 0.040000s user + 0.060000s system = 0.100000s CPU (263.4%)\n",
      "Kmers branching in  0.025672s wall, 0.850000s user + 0.020000s system = 0.870000s CPU (3388.9%)\n",
      "Kmer: 173 Graph size: 107041 Contigs in: 70\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.218538s wall, 2.390000s user + 0.000000s system = 2.390000s CPU (1093.6%)\n",
      "Kmers in multiple/single contigs: 0 53806\n",
      "Fragments before: 197 19628\n",
      "Fragments after: 157 19628\n",
      "New seeds: 0\n",
      "New seeds in  0.058698s wall, 2.000000s user + 0.010000s system = 2.010000s CPU (3424.3%)\n",
      "Fragments before: 127 14331\n",
      "Fragments after: 127 14331\n",
      "Connectors: 0 Extenders: 127\n",
      "Connections and extensions in  0.439891s wall, 4.490000s user + 0.070000s system = 4.560000s CPU (1036.6%)\n",
      "Contigs out: 67 Genome: 4837258 N50: 197750 L50: 9\n",
      "Assembled in  0.718937s wall, 8.890000s user + 0.080000s system = 8.970000s CPU (1247.7%)\n",
      "\n",
      "Seed kmers: 0\n",
      "Cleaned reads: 24440\n",
      "Reads for connection: 25864\n",
      "Internal reads: 1617\n",
      "Reads cleaned in  0.944530s wall, 8.060000s user + 0.390000s system = 8.450000s CPU (894.6%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 21\n",
      "Connected: 9635 ambiguously connected: 2548 from 12932 mate pairs\n",
      "Connect pairs in  1.886191s wall, 91.950000s user + 0.160000s system = 92.110000s CPU (4883.4%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 35\n",
      "Connected: 471 ambiguously connected: 2053 from 2548 mate pairs\n",
      "Connect pairs in  0.430554s wall, 15.230000s user + 0.350000s system = 15.580000s CPU (3618.6%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 51\n",
      "Connected: 194 ambiguously connected: 1831 from 2053 mate pairs\n",
      "Connect pairs in  0.352710s wall, 12.180000s user + 0.010000s system = 12.190000s CPU (3456.1%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 67\n",
      "Connected: 203 ambiguously connected: 1584 from 1831 mate pairs\n",
      "Connect pairs in  0.358134s wall, 15.270000s user + 0.020000s system = 15.290000s CPU (4269.4%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 81\n",
      "Connected: 108 ambiguously connected: 1435 from 1584 mate pairs\n",
      "Connect pairs in  0.267563s wall, 10.460000s user + 0.150000s system = 10.610000s CPU (3965.4%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 97\n",
      "Connected: 134 ambiguously connected: 1269 from 1435 mate pairs\n",
      "Connect pairs in  0.275442s wall, 10.360000s user + 0.000000s system = 10.360000s CPU (3761.2%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 111\n",
      "Connected: 57 ambiguously connected: 1188 from 1269 mate pairs\n",
      "Connect pairs in  0.266914s wall, 9.750000s user + 0.090000s system = 9.840000s CPU (3686.6%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 127\n",
      "Connected: 159 ambiguously connected: 973 from 1188 mate pairs\n",
      "Connect pairs in  0.282763s wall, 9.470000s user + 0.010000s system = 9.480000s CPU (3352.6%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 143\n",
      "Connected: 112 ambiguously connected: 837 from 973 mate pairs\n",
      "Connect pairs in  0.216545s wall, 8.430000s user + 0.010000s system = 8.440000s CPU (3897.6%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 157\n",
      "Connected: 63 ambiguously connected: 741 from 837 mate pairs\n",
      "Connect pairs in  0.218233s wall, 7.210000s user + 0.080000s system = 7.290000s CPU (3340.5%)\n",
      "\n",
      "Connecting mate pairs using kmer length: 173\n",
      "Connected: 57 ambiguously connected: 648 from 741 mate pairs\n",
      "Connect pairs in  0.215457s wall, 6.550000s user + 0.040000s system = 6.590000s CPU (3058.6%)\n",
      "Totally connected: 11193\n",
      "Added notconnected: 1295\n",
      "\n",
      "Kmer len: 215\n",
      "Raw kmers: 2409066 Memory needed (GB): 0.185016 Memory available (GB): 29.7224 1 cycle(s) will be performed\n",
      "Distinct kmers: 91339\n",
      "Kmer count in  0.116421s wall, 2.430000s user + 0.410000s system = 2.840000s CPU (2439.4%)\n",
      "Uniq kmers merging in  0.046087s wall, 0.060000s user + 0.030000s system = 0.090000s CPU (195.3%)\n",
      "Kmers branching in  0.021647s wall, 0.730000s user + 0.010000s system = 0.740000s CPU (3418.5%)\n",
      "Kmer: 215 Graph size: 91339 Contigs in: 67\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.252562s wall, 2.510000s user + 0.000000s system = 2.510000s CPU (993.8%)\n",
      "Kmers in multiple/single contigs: 0 55228\n",
      "Fragments before: 222 27101\n",
      "Fragments after: 151 27101\n",
      "New seeds: 0\n",
      "New seeds in  0.050571s wall, 1.650000s user + 0.010000s system = 1.660000s CPU (3282.5%)\n",
      "Fragments before: 123 19306\n",
      "Fragments after: 114 19306\n",
      "Connectors: 9 Extenders: 105\n",
      "Connections and extensions in  0.537774s wall, 4.240000s user + 0.170000s system = 4.410000s CPU (820.0%)\n",
      "Contigs out: 56 Genome: 4840295 N50: 231146 L50: 7\n",
      "Assembled in  0.842250s wall, 8.400000s user + 0.180000s system = 8.580000s CPU (1018.7%)\n",
      "\n",
      "\n",
      "Kmer len: 289\n",
      "Raw kmers: 1783850 Memory needed (GB): 0.188375 Memory available (GB): 29.7116 1 cycle(s) will be performed\n",
      "Distinct kmers: 90590\n",
      "Kmer count in  0.078291s wall, 2.180000s user + 0.060000s system = 2.240000s CPU (2861.1%)\n",
      "Uniq kmers merging in  0.040514s wall, 0.060000s user + 0.040000s system = 0.100000s CPU (246.8%)\n",
      "Kmers branching in  0.028405s wall, 0.880000s user + 0.000000s system = 0.880000s CPU (3098.1%)\n",
      "Kmer: 289 Graph size: 90590 Contigs in: 56\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.266794s wall, 2.630000s user + 0.000000s system = 2.630000s CPU (985.8%)\n",
      "Kmers in multiple/single contigs: 0 51840\n",
      "Fragments before: 167 27333\n",
      "Fragments after: 115 27333\n",
      "New seeds: 0\n",
      "New seeds in  0.060341s wall, 1.610000s user + 0.010000s system = 1.620000s CPU (2684.7%)\n",
      "Fragments before: 100 19281\n",
      "Fragments after: 91 19281\n",
      "Connectors: 9 Extenders: 82\n",
      "Connections and extensions in  0.674707s wall, 4.320000s user + 0.170000s system = 4.490000s CPU (665.5%)\n",
      "Contigs out: 45 Genome: 4843194 N50: 298365 L50: 5\n",
      "Assembled in  1.003285s wall, 8.560000s user + 0.180000s system = 8.740000s CPU (871.1%)\n",
      "\n",
      "\n",
      "Kmer len: 363\n",
      "Raw kmers: 1304455 Memory needed (GB): 0.162796 Memory available (GB): 29.6987 1 cycle(s) will be performed\n",
      "Distinct kmers: 89270\n",
      "Kmer count in  0.070424s wall, 1.830000s user + 0.070000s system = 1.900000s CPU (2697.9%)\n",
      "Uniq kmers merging in  0.047670s wall, 0.050000s user + 0.060000s system = 0.110000s CPU (230.8%)\n",
      "Kmers branching in  0.027068s wall, 1.040000s user + 0.010000s system = 1.050000s CPU (3879.2%)\n",
      "Kmer: 363 Graph size: 89270 Contigs in: 45\n",
      "Valley: 0\n",
      "\n",
      "Mark used kmers in  0.329231s wall, 2.800000s user + 0.020000s system = 2.820000s CPU (856.5%)\n",
      "Kmers in multiple/single contigs: 0 50178\n",
      "Fragments before: 160 28152\n",
      "Fragments after: 102 28152\n",
      "New seeds: 0\n",
      "New seeds in  0.060909s wall, 1.800000s user + 0.010000s system = 1.810000s CPU (2971.6%)\n",
      "Fragments before: 72 16045\n",
      "Fragments after: 70 16045\n",
      "Connectors: 2 Extenders: 68\n",
      "Connections and extensions in  0.841562s wall, 4.170000s user + 0.210000s system = 4.380000s CPU (520.5%)\n",
      "Contigs out: 38 Genome: 4842245 N50: 298538 L50: 4\n",
      "Assembled in  1.232267s wall, 8.770000s user + 0.240000s system = 9.010000s CPU (731.2%)\n",
      "\n",
      "DONE\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n skesa skesa --reads SRR3028792_1.fp.fastq.gz,SRR3028792_2.fp.fastq.gz --contigs_out SRR3028792.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb2af46a-d71f-4855-973b-3c0b091f1ffb",
   "metadata": {},
   "source": [
    "The meaning of the command-line arguments is as follows:\n",
    "\n",
    "* `--reads SRR3028792_1.fp.fastq.gz,SRR3028792_2.fp.fastq.gz`: The sequence reads. For paired-end data each pair of files needs to be separated by a comma `,`.\n",
    "* `--contigs_out SRR3028792.fasta`: The output file containing the longer contiguous sequences (contigs), in **fasta** format."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e8216ab-9196-4b42-bd30-f5d3e351b637",
   "metadata": {},
   "source": [
    "## 4.3. Examine output contigs (`seqkit`)\n",
    "\n",
    "We can examine the output contigs file quickly on the command-line using the [seqkit](https://bioinf.shenwei.me/seqkit/) command-line tool."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c8d4d1e-886f-438e-93fa-d55ddc0e3660",
   "metadata": {},
   "source": [
    "### 4.3.1. Install `seqkit`\n",
    "\n",
    "To install `seqkit` we can use conda."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "209934e2-189d-439a-a18a-a0a43686e712",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/seqkit\n",
      "\n",
      "  added / updated specs:\n",
      "    - seqkit\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  seqkit             bioconda/linux-64::seqkit-2.1.0-h9ee0642_0\n",
      "\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate seqkit\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda create -y -n seqkit seqkit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "28921195-5d86-472a-a40b-2a947a419439",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "simple statistics of FASTA/Q files\n",
      "\n",
      "Tips:\n",
      "  1. For lots of small files (especially on SDD), use big value of '-j' to\n",
      "     parallelize counting.\n",
      "\n",
      "Usage:\n",
      "  seqkit stats [flags]\n",
      "\n",
      "Aliases:\n",
      "  stats, stat\n",
      "\n",
      "Flags:\n",
      "  -a, --all                  all statistics, including quartiles of seq length, sum_gap, N50\n",
      "  -b, --basename             only output basename of files\n",
      "  -E, --fq-encoding string   fastq quality encoding. available values: 'sanger', 'solexa', 'illumina-1.3+', 'illumina-1.5+', 'illumina-1.8+'. (default \"sanger\")\n",
      "  -G, --gap-letters string   gap letters (default \"- .\")\n",
      "  -h, --help                 help for stats\n",
      "  -e, --skip-err             skip error, only show warning message\n",
      "  -i, --stdin-label string   label for replacing default \"-\" for stdin (default \"-\")\n",
      "  -T, --tabular              output in machine-friendly tabular format\n",
      "\n",
      "Global Flags:\n",
      "      --alphabet-guess-seq-length int   length of sequence prefix of the first FASTA record based on which seqkit guesses the sequence type (0 for whole seq) (default 10000)\n",
      "      --id-ncbi                         FASTA head is NCBI-style, e.g. >gi|110645304|ref|NC_002516.2| Pseud...\n",
      "      --id-regexp string                regular expression for parsing ID (default \"^(\\\\S+)\\\\s?\")\n",
      "      --infile-list string              file of input files list (one file per line), if given, they are appended to files from cli arguments\n",
      "  -w, --line-width int                  line width when outputting FASTA format (0 for no wrap) (default 60)\n",
      "  -o, --out-file string                 out file (\"-\" for stdout, suffix .gz for gzipped out) (default \"-\")\n",
      "      --quiet                           be quiet and do not show extra information\n",
      "  -t, --seq-type string                 sequence type (dna|rna|protein|unlimit|auto) (for auto, it automatically detect by the first sequence) (default \"auto\")\n",
      "  -j, --threads int                     number of CPUs. can also set with environment variable SEQKIT_THREADS) (default 4)\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n seqkit seqkit stats --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0772b751-56fe-4717-9cb1-9cd50f626beb",
   "metadata": {},
   "source": [
    "### 4.3.2. Print stats of contigs\n",
    "\n",
    "To print basic stats on the assembled genome (`SRR3028792.fasta`) we can run the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5e941745-93fc-4e6a-8daf-abf95a507a93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file              format  type  num_seqs    sum_len  min_len    avg_len    max_len   Q1      Q2       Q3  sum_gap      N50  Q20(%)  Q30(%)\n",
      "SRR3028792.fasta  FASTA   DNA         38  4,842,245      370  127,427.5  1,022,474  803  45,160  184,096        0  298,538       0       0\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n seqkit seqkit stats -a SRR3028792.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c48fa6be-b102-4785-9eaf-73ccb3a16cea",
   "metadata": {},
   "source": [
    "This gives us some basic information on our assembled genome. For example `sum_len` is the total length of all contigs (and so our genome, around 4.8 million bp). `num_seqs` is the number of contigs in our genome."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a15ee6b-9b19-47ec-a2a7-d9825edafaa9",
   "metadata": {},
   "source": [
    "# 5. Evaulate assembly quality (`quast`)\n",
    "\n",
    "While `seqkit stats` gives us a quick way to examine the assembled genome, it can also be nice to have a visual and interactive report. This is where [quast](https://github.com/ablab/quast) comes in, as it can produce an interactive report that you can open in your web browser.\n",
    "\n",
    "You can upload the `SRR3028792.fasta` file to the quast website at <http://cab.cc.spbu.ru/quast/>. However, we will go over how to run quast on the command-line (useful if you have many genomes you want to examine at once).\n",
    "\n",
    "## 5.1. Install `quast`\n",
    "\n",
    "The first step is to install a local version of `quast` using conda."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "603778c6-3265-432a-a70b-f1740251e69d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/quast\n",
      "\n",
      "  added / updated specs:\n",
      "    - quast\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
      "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
      "  blast              bioconda/linux-64::blast-2.12.0-pl5262h3289130_0\n",
      "  brotli             conda-forge/linux-64::brotli-1.0.9-h7f98852_6\n",
      "  brotli-bin         conda-forge/linux-64::brotli-bin-1.0.9-h7f98852_6\n",
      "  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4\n",
      "  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0\n",
      "  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0\n",
      "  certifi            conda-forge/linux-64::certifi-2021.10.8-py37h89c1867_1\n",
      "  circos             bioconda/noarch::circos-0.69.8-hdfd78af_1\n",
      "  curl               conda-forge/linux-64::curl-7.81.0-h494985f_0\n",
      "  cycler             conda-forge/noarch::cycler-0.11.0-pyhd8ed1ab_0\n",
      "  entrez-direct      bioconda/linux-64::entrez-direct-16.2-he881be0_0\n",
      "  expat              conda-forge/linux-64::expat-2.4.3-h9c3ff4c_0\n",
      "  fontconfig         conda-forge/linux-64::fontconfig-2.13.1-hba837de_1005\n",
      "  fonttools          conda-forge/linux-64::fonttools-4.28.5-py37h5e8e339_0\n",
      "  freetype           conda-forge/linux-64::freetype-2.10.4-h0708190_1\n",
      "  giflib             conda-forge/linux-64::giflib-5.2.1-h36c2ea0_2\n",
      "  glimmerhmm         bioconda/linux-64::glimmerhmm-3.0.4-pl526hf484d3e_3\n",
      "  icu                conda-forge/linux-64::icu-67.1-he1b5a44_0\n",
      "  jbig               conda-forge/linux-64::jbig-2.1-h7f98852_2003\n",
      "  joblib             conda-forge/noarch::joblib-1.1.0-pyhd8ed1ab_0\n",
      "  jpeg               conda-forge/linux-64::jpeg-9d-h36c2ea0_0\n",
      "  k8                 bioconda/linux-64::k8-0.2.5-h9a82719_1\n",
      "  kiwisolver         conda-forge/linux-64::kiwisolver-1.3.2-py37h2527ec5_1\n",
      "  krb5               conda-forge/linux-64::krb5-1.19.2-h48eae69_3\n",
      "  lcms2              conda-forge/linux-64::lcms2-2.12-hddcbb42_0\n",
      "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.36.1-hea4e1c9_2\n",
      "  lerc               conda-forge/linux-64::lerc-3.0-h9c3ff4c_0\n",
      "  libblas            conda-forge/linux-64::libblas-3.9.0-13_linux64_openblas\n",
      "  libbrotlicommon    conda-forge/linux-64::libbrotlicommon-1.0.9-h7f98852_6\n",
      "  libbrotlidec       conda-forge/linux-64::libbrotlidec-1.0.9-h7f98852_6\n",
      "  libbrotlienc       conda-forge/linux-64::libbrotlienc-1.0.9-h7f98852_6\n",
      "  libcblas           conda-forge/linux-64::libcblas-3.9.0-13_linux64_openblas\n",
      "  libcurl            conda-forge/linux-64::libcurl-7.81.0-h494985f_0\n",
      "  libdeflate         conda-forge/linux-64::libdeflate-1.8-h7f98852_0\n",
      "  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2\n",
      "  libev              conda-forge/linux-64::libev-4.33-h516909a_1\n",
      "  libffi             conda-forge/linux-64::libffi-3.4.2-h7f98852_5\n",
      "  libgcc             conda-forge/linux-64::libgcc-7.2.0-h69d50b8_2\n",
      "  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11\n",
      "  libgd              conda-forge/linux-64::libgd-2.2.5-h6e6070c_1008\n",
      "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-11.2.0-h69a702a_11\n",
      "  libgfortran5       conda-forge/linux-64::libgfortran5-11.2.0-h5c6108e_11\n",
      "  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11\n",
      "  libiconv           conda-forge/linux-64::libiconv-1.16-h516909a_0\n",
      "  liblapack          conda-forge/linux-64::liblapack-3.9.0-13_linux64_openblas\n",
      "  libnghttp2         conda-forge/linux-64::libnghttp2-1.43.0-ha19adfc_1\n",
      "  libnsl             conda-forge/linux-64::libnsl-2.0.0-h7f98852_0\n",
      "  libopenblas        conda-forge/linux-64::libopenblas-0.3.18-pthreads_h8fe5266_0\n",
      "  libpng             conda-forge/linux-64::libpng-1.6.37-h21135ba_2\n",
      "  libssh2            conda-forge/linux-64::libssh2-1.10.0-ha35d2d1_2\n",
      "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_11\n",
      "  libtiff            conda-forge/linux-64::libtiff-4.3.0-h6f004c6_2\n",
      "  libuuid            conda-forge/linux-64::libuuid-2.32.1-h7f98852_1000\n",
      "  libwebp            conda-forge/linux-64::libwebp-1.2.1-h3452ae3_0\n",
      "  libwebp-base       conda-forge/linux-64::libwebp-base-1.2.1-h7f98852_0\n",
      "  libxml2            conda-forge/linux-64::libxml2-2.9.10-h68273f3_2\n",
      "  libzlib            conda-forge/linux-64::libzlib-1.2.11-h36c2ea0_1013\n",
      "  lz4-c              conda-forge/linux-64::lz4-c-1.9.3-h9c3ff4c_1\n",
      "  matplotlib-base    conda-forge/linux-64::matplotlib-base-3.5.1-py37h1058ff1_0\n",
      "  minimap2           bioconda/linux-64::minimap2-2.24-h5bf99c6_0\n",
      "  munkres            bioconda/noarch::munkres-1.0.7-py_1\n",
      "  ncurses            conda-forge/linux-64::ncurses-6.3-h9c3ff4c_0\n",
      "  numpy              conda-forge/linux-64::numpy-1.21.5-py37hf2998dd_0\n",
      "  olefile            conda-forge/noarch::olefile-0.46-pyh9f0ad1d_1\n",
      "  openjdk            conda-forge/linux-64::openjdk-10.0.2-h14c3975_1015\n",
      "  openjpeg           conda-forge/linux-64::openjpeg-2.4.0-hb52868f_1\n",
      "  openssl            conda-forge/linux-64::openssl-3.0.0-h7f98852_2\n",
      "  packaging          conda-forge/noarch::packaging-21.3-pyhd8ed1ab_0\n",
      "  pcre               conda-forge/linux-64::pcre-8.45-h9c3ff4c_0\n",
      "  perl               conda-forge/linux-64::perl-5.26.2-h36c2ea0_1008\n",
      "  perl-archive-tar   bioconda/linux-64::perl-archive-tar-2.32-pl526_0\n",
      "  perl-autoloader    bioconda/linux-64::perl-autoloader-5.74-pl526_2\n",
      "  perl-carp          bioconda/linux-64::perl-carp-1.38-pl526_3\n",
      "  perl-clone         bioconda/linux-64::perl-clone-0.42-pl526h516909a_0\n",
      "  perl-common-sense  bioconda/linux-64::perl-common-sense-3.74-pl526_2\n",
      "  perl-compress-raw~ bioconda/linux-64::perl-compress-raw-bzip2-2.087-pl526he1b5a44_0\n",
      "  perl-compress-raw~ bioconda/linux-64::perl-compress-raw-zlib-2.087-pl526hc9558a2_0\n",
      "  perl-config-gener~ bioconda/linux-64::perl-config-general-2.63-pl526_0\n",
      "  perl-digest-perl-~ bioconda/linux-64::perl-digest-perl-md5-1.9-pl526_1\n",
      "  perl-dynaloader    bioconda/linux-64::perl-dynaloader-1.25-pl526_1\n",
      "  perl-exporter      bioconda/linux-64::perl-exporter-5.72-pl526_1\n",
      "  perl-exporter-tiny bioconda/linux-64::perl-exporter-tiny-1.002001-pl526_0\n",
      "  perl-extutils-mak~ bioconda/linux-64::perl-extutils-makemaker-7.36-pl526_1\n",
      "  perl-font-ttf      bioconda/linux-64::perl-font-ttf-1.06-pl526_0\n",
      "  perl-gd            bioconda/linux-64::perl-gd-2.71-pl526he860b03_0\n",
      "  perl-io-compress   bioconda/linux-64::perl-io-compress-2.087-pl526he1b5a44_0\n",
      "  perl-io-string     bioconda/linux-64::perl-io-string-1.08-pl526_3\n",
      "  perl-io-zlib       bioconda/linux-64::perl-io-zlib-1.10-pl526_2\n",
      "  perl-json          bioconda/linux-64::perl-json-4.02-pl526_0\n",
      "  perl-json-xs       bioconda/linux-64::perl-json-xs-2.34-pl526h6bb024c_3\n",
      "  perl-list-moreuti~ bioconda/linux-64::perl-list-moreutils-0.428-pl526_1\n",
      "  perl-list-moreuti~ bioconda/linux-64::perl-list-moreutils-xs-0.428-pl526_0\n",
      "  perl-math-bezier   bioconda/linux-64::perl-math-bezier-0.01-pl526_1\n",
      "  perl-math-round    bioconda/linux-64::perl-math-round-0.07-pl526_1\n",
      "  perl-math-vecstat  bioconda/linux-64::perl-math-vecstat-0.08-pl526_1\n",
      "  perl-module-imple~ bioconda/linux-64::perl-module-implementation-0.09-pl526_2\n",
      "  perl-module-runti~ bioconda/linux-64::perl-module-runtime-0.016-pl526_1\n",
      "  perl-number-format bioconda/linux-64::perl-number-format-1.75-pl526_3\n",
      "  perl-params-valid~ bioconda/linux-64::perl-params-validate-1.29-pl526h14c3975_1\n",
      "  perl-pathtools     bioconda/linux-64::perl-pathtools-3.75-pl526h14c3975_1\n",
      "  perl-readonly      bioconda/linux-64::perl-readonly-2.05-pl526_0\n",
      "  perl-regexp-common bioconda/linux-64::perl-regexp-common-2017060201-pl526_0\n",
      "  perl-scalar-list-~ bioconda/linux-64::perl-scalar-list-utils-1.52-pl526h516909a_0\n",
      "  perl-set-intspan   bioconda/linux-64::perl-set-intspan-1.19-pl526_1\n",
      "  perl-statistics-b~ bioconda/linux-64::perl-statistics-basic-1.6611-pl526_2\n",
      "  perl-svg           bioconda/linux-64::perl-svg-2.84-pl526_0\n",
      "  perl-text-format   bioconda/linux-64::perl-text-format-0.59-pl526_2\n",
      "  perl-threaded      bioconda/noarch::perl-threaded-5.32.1-hdfd78af_1\n",
      "  perl-time-hires    bioconda/linux-64::perl-time-hires-1.9760-pl526h14c3975_1\n",
      "  perl-try-tiny      bioconda/linux-64::perl-try-tiny-0.30-pl526_1\n",
      "  perl-types-serial~ bioconda/linux-64::perl-types-serialiser-1.0-pl526_2\n",
      "  perl-xml-parser    bioconda/linux-64::perl-xml-parser-2.44-3\n",
      "  perl-xsloader      bioconda/linux-64::perl-xsloader-0.24-pl526_0\n",
      "  pillow             conda-forge/linux-64::pillow-8.4.0-py37h0f21c89_0\n",
      "  pip                conda-forge/noarch::pip-21.3.1-pyhd8ed1ab_0\n",
      "  pyparsing          conda-forge/noarch::pyparsing-3.0.7-pyhd8ed1ab_0\n",
      "  python             conda-forge/linux-64::python-3.7.12-hf930737_100_cpython\n",
      "  python-dateutil    conda-forge/noarch::python-dateutil-2.8.2-pyhd8ed1ab_0\n",
      "  python_abi         conda-forge/linux-64::python_abi-3.7-2_cp37m\n",
      "  quast              bioconda/linux-64::quast-5.0.2-py37pl5262hfecc14a_5\n",
      "  readline           conda-forge/linux-64::readline-8.1-h46c0cb4_0\n",
      "  setuptools         conda-forge/linux-64::setuptools-60.5.0-py37h89c1867_0\n",
      "  simplejson         conda-forge/linux-64::simplejson-3.17.6-py37h5e8e339_0\n",
      "  six                conda-forge/noarch::six-1.16.0-pyh6c4a22f_0\n",
      "  sqlite             conda-forge/linux-64::sqlite-3.37.0-h9cd32fc_0\n",
      "  tk                 conda-forge/linux-64::tk-8.6.11-h27826a3_1\n",
      "  unicodedata2       conda-forge/linux-64::unicodedata2-14.0.0-py37h5e8e339_0\n",
      "  wheel              conda-forge/noarch::wheel-0.37.1-pyhd8ed1ab_0\n",
      "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
      "  zlib               conda-forge/linux-64::zlib-1.2.11-h36c2ea0_1013\n",
      "  zstd               conda-forge/linux-64::zstd-1.5.2-ha95c52a_0\n",
      "\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: \\ The default QUAST package does not include:\n",
      "* GRIDSS (needed for structural variants detection)\n",
      "* SILVA 16S rRNA database (needed for reference genome detection in metagenomic datasets)\n",
      "* BUSCO tools and databases (needed for searching BUSCO genes) -- works in Linux only!\n",
      "\n",
      "To be able to use those, please run\n",
      "    quast-download-gridss\n",
      "    quast-download-silva\n",
      "    quast-download-busco\n",
      "\n",
      "done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate quast\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda create -y -n quast quast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5640c41e-c3f0-4d64-9db1-2cb35adea2f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QUAST: Quality Assessment Tool for Genome Assemblies\n",
      "Version: 5.0.2\n",
      "\n",
      "Usage: python /home/CSCScience.ca/apetkau/miniconda3/envs/quast/bin/quast [options] <files_with_contigs>\n",
      "\n",
      "Options:\n",
      "-o  --output-dir  <dirname>       Directory to store all result files [default: quast_results/results_<datetime>]\n",
      "-r                <filename>      Reference genome file\n",
      "-g  --features [type:]<filename>  File with genomic feature coordinates in the reference (GFF, BED, NCBI or TXT)\n",
      "                                  Optional 'type' can be specified for extracting only a specific feature type from GFF\n",
      "-m  --min-contig  <int>           Lower threshold for contig length [default: 500]\n",
      "-t  --threads     <int>           Maximum number of threads [default: 25% of CPUs]\n",
      "\n",
      "Advanced options:\n",
      "-s  --split-scaffolds                 Split assemblies by continuous fragments of N's and add such \"contigs\" to the comparison\n",
      "-l  --labels \"label, label, ...\"      Names of assemblies to use in reports, comma-separated. If contain spaces, use quotes\n",
      "-L                                    Take assembly names from their parent directory names\n",
      "-e  --eukaryote                       Genome is eukaryotic (primarily affects gene prediction)\n",
      "    --fungus                          Genome is fungal (primarily affects gene prediction)\n",
      "    --large                           Use optimal parameters for evaluation of large genomes\n",
      "                                      In particular, imposes '-e -m 3000 -i 500 -x 7000' (can be overridden manually)\n",
      "-k  --k-mer-stats                     Compute k-mer-based quality metrics (recommended for large genomes)\n",
      "                                      This may significantly increase memory and time consumption on large genomes\n",
      "    --k-mer-size                      Size of k used in --k-mer-stats [default: 101]\n",
      "    --circos                          Draw Circos plot\n",
      "-f  --gene-finding                    Predict genes using GeneMarkS (prokaryotes, default) or GeneMark-ES (eukaryotes, use --eukaryote)\n",
      "    --mgm                             Use MetaGeneMark for gene prediction (instead of the default finder, see above)\n",
      "    --glimmer                         Use GlimmerHMM for gene prediction (instead of the default finder, see above)\n",
      "    --gene-thresholds <int,int,...>   Comma-separated list of threshold lengths of genes to search with Gene Finding module\n",
      "                                      [default: 0,300,1500,3000]\n",
      "    --rna-finding                     Predict ribosomal RNA genes using Barrnap\n",
      "-b  --conserved-genes-finding         Count conserved orthologs using BUSCO (only on Linux)\n",
      "    --operons  <filename>             File with operon coordinates in the reference (GFF, BED, NCBI or TXT)\n",
      "    --est-ref-size <int>              Estimated reference size (for computing NGx metrics without a reference)\n",
      "    --contig-thresholds <int,int,...> Comma-separated list of contig length thresholds [default: 0,1000,5000,10000,25000,50000]\n",
      "-u  --use-all-alignments              Compute genome fraction, # genes, # operons in QUAST v1.* style.\n",
      "                                      By default, QUAST filters Minimap's alignments to keep only best ones\n",
      "-i  --min-alignment <int>             The minimum alignment length [default: 65]\n",
      "    --min-identity <float>            The minimum alignment identity (80.0, 100.0) [default: 95.0]\n",
      "-a  --ambiguity-usage <none|one|all>  Use none, one, or all alignments of a contig when all of them\n",
      "                                      are almost equally good (see --ambiguity-score) [default: one]\n",
      "    --ambiguity-score <float>         Score S for defining equally good alignments of a single contig. All alignments are sorted \n",
      "                                      by decreasing LEN * IDY% value. All alignments with LEN * IDY% < S * best(LEN * IDY%) are \n",
      "                                      discarded. S should be between 0.8 and 1.0 [default: 0.99]\n",
      "    --strict-NA                       Break contigs in any misassembly event when compute NAx and NGAx.\n",
      "                                      By default, QUAST breaks contigs only by extensive misassemblies (not local ones)\n",
      "-x  --extensive-mis-size  <int>       Lower threshold for extensive misassembly size. All relocations with inconsistency\n",
      "                                      less than extensive-mis-size are counted as local misassemblies [default: 1000]\n",
      "    --scaffold-gap-max-size  <int>    Max allowed scaffold gap length difference. All relocations with inconsistency\n",
      "                                      less than scaffold-gap-size are counted as scaffold gap misassemblies [default: 10000]\n",
      "    --unaligned-part-size  <int>      Lower threshold for detecting partially unaligned contigs. Such contig should have\n",
      "                                      at least one unaligned fragment >= the threshold [default: 500]\n",
      "    --skip-unaligned-mis-contigs      Do not distinguish contigs with >= 50% unaligned bases as a separate group\n",
      "                                      By default, QUAST does not count misassemblies in them\n",
      "    --fragmented                      Reference genome may be fragmented into small pieces (e.g. scaffolded reference) \n",
      "    --fragmented-max-indent  <int>    Mark translocation as fake if both alignments are located no further than N bases \n",
      "                                      from the ends of the reference fragments [default: 85]\n",
      "                                      Requires --fragmented option\n",
      "    --upper-bound-assembly            Simulate upper bound assembly based on the reference genome and reads\n",
      "    --upper-bound-min-con  <int>      Minimal number of 'connecting reads' needed for joining upper bound contigs into a scaffold\n",
      "                                      [default: 2 for mate-pairs and 1 for long reads]\n",
      "    --est-insert-size  <int>          Use provided insert size in upper bound assembly simulation [default: auto detect from reads or 255]\n",
      "    --plots-format  <str>             Save plots in specified format [default: pdf].\n",
      "                                      Supported formats: emf, eps, pdf, png, ps, raw, rgba, svg, svgz\n",
      "    --memory-efficient                Run everything using one thread, separately per each assembly.\n",
      "                                      This may significantly reduce memory consumption on large genomes\n",
      "    --space-efficient                 Create only reports and plots files. Aux files including .stdout, .stderr, .coords will not be created.\n",
      "                                      This may significantly reduce space consumption on large genomes. Icarus viewers also will not be built\n",
      "-1  --pe1     <filename>              File with forward paired-end reads (in FASTQ format, may be gzipped)\n",
      "-2  --pe2     <filename>              File with reverse paired-end reads (in FASTQ format, may be gzipped)\n",
      "    --pe12    <filename>              File with interlaced forward and reverse paired-end reads. (in FASTQ format, may be gzipped)\n",
      "    --mp1     <filename>              File with forward mate-pair reads (in FASTQ format, may be gzipped)\n",
      "    --mp2     <filename>              File with reverse mate-pair reads (in FASTQ format, may be gzipped)\n",
      "    --mp12    <filename>              File with interlaced forward and reverse mate-pair reads (in FASTQ format, may be gzipped)\n",
      "    --single  <filename>              File with unpaired short reads (in FASTQ format, may be gzipped)\n",
      "    --pacbio     <filename>           File with PacBio reads (in FASTQ format, may be gzipped)\n",
      "    --nanopore   <filename>           File with Oxford Nanopore reads (in FASTQ format, may be gzipped)\n",
      "    --ref-sam <filename>              SAM alignment file obtained by aligning reads to reference genome file\n",
      "    --ref-bam <filename>              BAM alignment file obtained by aligning reads to reference genome file\n",
      "    --sam     <filename,filename,...> Comma-separated list of SAM alignment files obtained by aligning reads to assemblies\n",
      "                                      (use the same order as for files with contigs)\n",
      "    --bam     <filename,filename,...> Comma-separated list of BAM alignment files obtained by aligning reads to assemblies\n",
      "                                      (use the same order as for files with contigs)\n",
      "                                      Reads (or SAM/BAM file) are used for structural variation detection and\n",
      "                                      coverage histogram building in Icarus\n",
      "    --sv-bedpe  <filename>            File with structural variations (in BEDPE format)\n",
      "\n",
      "Speedup options:\n",
      "    --no-check                        Do not check and correct input fasta files. Use at your own risk (see manual)\n",
      "    --no-plots                        Do not draw plots\n",
      "    --no-html                         Do not build html reports and Icarus viewers\n",
      "    --no-icarus                       Do not build Icarus viewers\n",
      "    --no-snps                         Do not report SNPs (may significantly reduce memory consumption on large genomes)\n",
      "    --no-gc                           Do not compute GC% and GC-distribution\n",
      "    --no-sv                           Do not run structural variation detection (make sense only if reads are specified)\n",
      "    --no-gzip                         Do not compress large output files\n",
      "    --no-read-stats                   Do not align reads to assemblies\n",
      "                                      Reads will be aligned to reference and used for coverage analysis,\n",
      "                                      upper bound assembly simulation, and structural variation detection.\n",
      "                                      Use this option if you do not need read statistics for assemblies.\n",
      "    --fast                            A combination of all speedup options except --no-check\n",
      "\n",
      "Other:\n",
      "    --silent                          Do not print detailed information about each step to stdout (log file is not affected)\n",
      "    --test                            Run QUAST on the data from the test_data folder, output to quast_test_output\n",
      "    --test-sv                         Run QUAST with structural variants detection on the data from the test_data folder,\n",
      "                                      output to quast_test_output\n",
      "-h  --help                            Print full usage message\n",
      "-v  --version                         Print version\n",
      "\n",
      "Online QUAST manual is available at http://quast.sf.net/manual\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n quast quast --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acc66f85-118e-4cf8-92c1-09070f568cb5",
   "metadata": {},
   "source": [
    "## 5.2. Create a quality report with `quast`\n",
    "\n",
    "Now we can run quast as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "32e19c56-423d-4226-bb11-0b4b52b6bf57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/miniconda3/envs/quast/bin/quast SRR3028792.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-161-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.7.12\n",
      "  CPUs number: 56\n",
      "\n",
      "Started: 2022-01-21 13:07:14\n",
      "\n",
      "Logging to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/quast.log\n",
      "NOTICE: Maximum number of threads is set to 14 (use --threads option to set it manually)\n",
      "\n",
      "CWD: /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial\n",
      "Main parameters: \n",
      "  MODE: default, threads: 14, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: one, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  SRR3028792.fasta ==> SRR3028792\n",
      "\n",
      "2022-01-21 13:07:15\n",
      "Running Basic statistics processor...\n",
      "  Contig files: \n",
      "    SRR3028792\n",
      "  Calculating N50 and L50...\n",
      "    SRR3028792, N50 = 298538, L50 = 4, Total length = 4839502, GC % = 52.04, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/basic_stats/Nx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/basic_stats/GC_content_plot.pdf\n",
      "  Drawing SRR3028792 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/basic_stats/SRR3028792_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2022-01-21 13:07:16\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2022-01-21 13:07:17\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/report.html\n",
      "  PDF version (tables and plots) is saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/report.pdf\n",
      "  Icarus (contig browser) is saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/icarus.html\n",
      "  Log is saved to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_21_13_07_14/quast.log\n",
      "\n",
      "Finished: 2022-01-21 13:07:17\n",
      "Elapsed time: 0:00:02.080033\n",
      "NOTICEs: 2; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n quast quast SRR3028792.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f30493ca-4338-4630-a5ca-3ee82da27ae5",
   "metadata": {},
   "source": [
    "If you have more than one genome to evaluate you can run list them all on the command line `quast genome1.fasta genome2.fasta ...` or just use `quast *.fasta`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2ec972d-5049-4529-933a-5b0445ccf568",
   "metadata": {},
   "source": [
    "### 5.2.1 Zip up report for download\n",
    "\n",
    "The output of quast can be found in the directory `quast_results/latest/report.html` and you can open this in a web browser. However, if you are running this using Jupyter then we have to download the entire directory to your local computer first before viewing.\n",
    "\n",
    "We can download the entire `quast_results/` directory by first zipping up the quast report. You can create a `*.zip` archive from the command-line with the `zip` command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "87bd5bcd-7871-43d0-bac7-6fd72cdc5f51",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  adding: quast_results/ (stored 0%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/ (stored 0%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/report.tsv (deflated 57%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/quast.log (deflated 71%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/transposed_report.tsv (deflated 58%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/report.pdf (deflated 27%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/icarus.html (deflated 85%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/report.tex (deflated 64%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/transposed_report.txt (deflated 67%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/transposed_report.tex (deflated 60%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/report.html (deflated 73%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/basic_stats/ (stored 0%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/basic_stats/SRR3028792_GC_content_plot.pdf (deflated 29%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/basic_stats/cumulative_plot.pdf (deflated 30%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/basic_stats/GC_content_plot.pdf (deflated 28%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/basic_stats/Nx_plot.pdf (deflated 31%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/icarus_viewers/ (stored 0%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/icarus_viewers/contig_size_viewer.html (deflated 79%)\n",
      "  adding: quast_results/results_2022_01_21_13_07_14/report.txt (deflated 67%)\n",
      "  adding: quast_results/latest/ (stored 0%)\n",
      "  adding: quast_results/latest/report.tsv (deflated 57%)\n",
      "  adding: quast_results/latest/quast.log (deflated 71%)\n",
      "  adding: quast_results/latest/transposed_report.tsv (deflated 58%)\n",
      "  adding: quast_results/latest/report.pdf (deflated 27%)\n",
      "  adding: quast_results/latest/icarus.html (deflated 85%)\n",
      "  adding: quast_results/latest/report.tex (deflated 64%)\n",
      "  adding: quast_results/latest/transposed_report.txt (deflated 67%)\n",
      "  adding: quast_results/latest/transposed_report.tex (deflated 60%)\n",
      "  adding: quast_results/latest/report.html (deflated 73%)\n",
      "  adding: quast_results/latest/basic_stats/ (stored 0%)\n",
      "  adding: quast_results/latest/basic_stats/SRR3028792_GC_content_plot.pdf (deflated 29%)\n",
      "  adding: quast_results/latest/basic_stats/cumulative_plot.pdf (deflated 30%)\n",
      "  adding: quast_results/latest/basic_stats/GC_content_plot.pdf (deflated 28%)\n",
      "  adding: quast_results/latest/basic_stats/Nx_plot.pdf (deflated 31%)\n",
      "  adding: quast_results/latest/icarus_viewers/ (stored 0%)\n",
      "  adding: quast_results/latest/icarus_viewers/contig_size_viewer.html (deflated 79%)\n",
      "  adding: quast_results/latest/report.txt (deflated 67%)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "zip -r quast.zip quast_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb8435a0-32c7-4303-a685-adcf6b9b9d63",
   "metadata": {},
   "source": [
    "Now you can download the zip file by right-clicking `quast.zip` and selecting **Download**.\n",
    "\n",
    "![jupyter-quast-zip.png](../images/jupyter-quast-zip.png)\n",
    "\n",
    "Next, you can unzip the `quast.zip` file on your local computer and open `quast_report/latest/report.html` in a web browser. ou should end up with something that looks like:\n",
    "\n",
    "![quast-report.png](../images/quast-report.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828277fb-a6e3-4cc5-8053-183b415ec0e9",
   "metadata": {},
   "source": [
    "# 6. Where to go from here\n",
    "\n",
    "An assembled genome file (`SRR3028792.fasta`) often forms the basis for other downstream analysis. Here are a few examples you can try on your own."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe5482d-fa0a-4f5f-b0be-1d2e18c9c0d0",
   "metadata": {},
   "source": [
    "## 6.1. Anti-microbial resistance\n",
    "\n",
    "Try uploading `SRR3028792.fasta` to CARD/RGI (which will identify genes associated with antimicrobial resistance): https://card.mcmaster.ca/analyze/rgi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b60ed60-d616-4acf-8ada-98d29f3ebda2",
   "metadata": {},
   "source": [
    "## 6.2. Multi-locus sequence typing\n",
    "\n",
    "Try downloading and running [mlst](https://github.com/tseemann/mlst) on the `SRR3028792.fasta` file, which will give you the [Multi-locus sequence type](https://en.wikipedia.org/wiki/Multilocus_sequence_typing) of the genome. You can do this with the following commands.\n",
    "\n",
    "First we will install [mamba](https://mamba.readthedocs.io/en/latest/), which is a faster version of `conda`. For most cases using the `conda` command is okay, but sometimes conda is slow at installing tools and `mamba` is meant to speed things up. I found that `conda` was too slow to install `mlst`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "966d3a32-c930-49ea-ab85-c432a72b0447",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n",
      "(base) \n"
     ]
    }
   ],
   "source": [
    "conda install -y mamba"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6a1212-2568-45e1-a1e0-e0c0799d0907",
   "metadata": {},
   "source": [
    "Next we can use `mamba` to install `mlst`. `mamba` is meant to replace `conda` so you can run the same `conda create` command as you're used to be replace it with `mamba create`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d24e5019-83e9-4c5c-8340-3fd9b7bd0d51",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Package                                       Version  Build             Channel                    Size\n",
      "\n",
      "  Install:\n",
      "\n",
      "\n",
      "\u001b[32m  + _libgcc_mutex                        \u001b[00m           0.1  conda_forge       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + _openmp_mutex                        \u001b[00m           4.5  1_gnu             conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + any2fasta                            \u001b[00m         0.4.2  hdfd78af_3        bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + blast                                \u001b[00m        2.12.0  pl5262h3289130_0  bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + bzip2                                \u001b[00m         1.0.8  h7f98852_4        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + c-ares                               \u001b[00m        1.18.1  h7f98852_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + ca-certificates                      \u001b[00m     2021.10.8  ha878542_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + cairo                                \u001b[00m        1.16.0  h6cf1ce9_1008     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + certifi                              \u001b[00m    2019.11.28  py27h8c360ce_1    conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + clustalw                             \u001b[00m           2.1  h7d875b9_6        bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + curl                                 \u001b[00m        7.81.0  h494985f_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + entrez-direct                        \u001b[00m          16.2  he881be0_0        bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + expat                                \u001b[00m        2.2.10  h9c3ff4c_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + fontconfig                           \u001b[00m        2.13.1  hba837de_1005     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + freetype                             \u001b[00m        2.10.4  h0708190_1        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + fribidi                              \u001b[00m        1.0.10  h36c2ea0_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + gettext                              \u001b[00m      0.19.8.1  h0b5b191_1005     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + giflib                               \u001b[00m         5.2.1  h36c2ea0_2        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + graphite2                            \u001b[00m        1.3.13  h58526e2_1001     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + graphviz                             \u001b[00m        2.42.3  h0511662_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + harfbuzz                             \u001b[00m         2.9.1  h83ec7ef_1        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + icu                                  \u001b[00m          68.2  h9c3ff4c_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + jpeg                                 \u001b[00m            9d  h36c2ea0_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + krb5                                 \u001b[00m        1.19.2  h48eae69_3        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libcurl                              \u001b[00m        7.81.0  h494985f_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libdb                                \u001b[00m        6.2.32  h9c3ff4c_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libedit                              \u001b[00m  3.1.20191231  he28a2e2_2        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libev                                \u001b[00m          4.33  h516909a_1        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libffi                               \u001b[00m           3.3  h58526e2_2        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libgcc                               \u001b[00m         7.2.0  h69d50b8_2        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libgcc-ng                            \u001b[00m        11.2.0  h1d223b6_11       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libgd                                \u001b[00m         2.2.5  h8e06009_4        pkgs/main/linux-64  \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libglib                              \u001b[00m        2.68.4  h3e27bee_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libgomp                              \u001b[00m        11.2.0  h1d223b6_11       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libiconv                             \u001b[00m          1.16  h516909a_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libnghttp2                           \u001b[00m        1.43.0  ha19adfc_1        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libpng                               \u001b[00m        1.6.37  h21135ba_2        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libssh2                              \u001b[00m        1.10.0  ha35d2d1_2        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libstdcxx-ng                         \u001b[00m        11.2.0  he4da1e4_11       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libtiff                              \u001b[00m         4.1.0  hc3755c2_3        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libtool                              \u001b[00m         2.4.6  h9c3ff4c_1008     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libuuid                              \u001b[00m        2.32.1  h7f98852_1000     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libwebp                              \u001b[00m         1.0.2  h56121f0_5        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libxcb                               \u001b[00m          1.13  h7f98852_1004     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libxml2                              \u001b[00m        2.9.12  h72842e0_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libxslt                              \u001b[00m        1.1.33  h15afd5d_2        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libzlib                              \u001b[00m        1.2.11  h36c2ea0_1013     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + lz4-c                                \u001b[00m         1.9.3  h9c3ff4c_1        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + mlst                                 \u001b[00m        2.19.0  hdfd78af_1        bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + ncurses                              \u001b[00m           6.3  h9c3ff4c_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + openssl                              \u001b[00m         3.0.0  h7f98852_2        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + paml                                 \u001b[00m           4.9  h779adbc_6        bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pango                                \u001b[00m        1.42.4  h69149e4_5        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pcre                                 \u001b[00m          8.45  h9c3ff4c_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl                                 \u001b[00m        5.26.2  h36c2ea0_1008     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-aceperl                         \u001b[00m          1.92  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-algorithm-diff                  \u001b[00m        1.1903  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-algorithm-munkres               \u001b[00m          0.08  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-apache-test                     \u001b[00m          1.40  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-app-cpanminus                   \u001b[00m        1.7044  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-appconfig                       \u001b[00m          1.71  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-archive-tar                     \u001b[00m          2.32  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-array-compare                   \u001b[00m         3.0.1  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-autoloader                      \u001b[00m          5.74  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-base                            \u001b[00m          2.23  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bio-asn1-entrezgene             \u001b[00m          1.73  pl5262hdfd78af_2  bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bio-coordinate                  \u001b[00m      1.007001  pl5262hdfd78af_2  bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bio-featureio                   \u001b[00m       1.6.905  pl5262hdfd78af_3  bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bio-phylo                       \u001b[00m          0.58  pl5262hdfd78af_3  bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bio-samtools                    \u001b[00m          1.43  pl526h1341992_1   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bio-tools-phylo-paml            \u001b[00m         1.7.3  pl5262hdfd78af_2  bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bio-tools-run-alignment-clustalw\u001b[00m         1.7.4  pl5262hdfd78af_2  bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bio-tools-run-alignment-tcoffee \u001b[00m         1.7.4  pl5262hdfd78af_3  bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bioperl                         \u001b[00m         1.7.2  pl526_11          bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bioperl-core                    \u001b[00m      1.007002  pl5262hdfd78af_3  bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-bioperl-run                     \u001b[00m      1.007002  pl5262hdfd78af_5  bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-business-isbn                   \u001b[00m         3.004  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-business-isbn-data              \u001b[00m  20140910.003  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-cache-cache                     \u001b[00m          1.08  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-capture-tiny                    \u001b[00m          0.48  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-carp                            \u001b[00m          1.38  pl526_3           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-cgi                             \u001b[00m          4.44  pl526h14c3975_1   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-class-data-inheritable          \u001b[00m          0.08  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-class-inspector                 \u001b[00m          1.34  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-class-load                      \u001b[00m          0.25  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-class-load-xs                   \u001b[00m          0.10  pl526h6bb024c_2   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-class-method-modifiers          \u001b[00m          2.12  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-clone                           \u001b[00m          0.42  pl526h516909a_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-common-sense                    \u001b[00m          3.74  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-compress-raw-bzip2              \u001b[00m         2.087  pl526he1b5a44_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-compress-raw-zlib               \u001b[00m         2.087  pl526hc9558a2_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-constant                        \u001b[00m          1.33  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-convert-binary-c                \u001b[00m          0.78  pl526h6bb024c_3   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-convert-binhex                  \u001b[00m         1.125  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-crypt-rc4                       \u001b[00m          2.02  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-data-dumper                     \u001b[00m         2.173  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-data-optlist                    \u001b[00m         0.110  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-data-stag                       \u001b[00m          0.14  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-date-format                     \u001b[00m          2.30  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-db-file                         \u001b[00m         1.855  pl526h516909a_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-dbd-sqlite                      \u001b[00m          1.64  pl526h516909a_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-dbi                             \u001b[00m         1.642  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-devel-globaldestruction         \u001b[00m          0.14  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-devel-overloadinfo              \u001b[00m         0.005  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-devel-stacktrace                \u001b[00m          2.04  pl526_0           bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-digest-hmac                     \u001b[00m          1.03  pl526_3           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-digest-md5                      \u001b[00m          2.55  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-digest-perl-md5                 \u001b[00m           1.9  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-digest-sha1                     \u001b[00m          2.13  pl526h6bb024c_1   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-dist-checkconflicts             \u001b[00m          0.11  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-dynaloader                      \u001b[00m          1.25  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-email-date-format               \u001b[00m         1.005  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-encode                          \u001b[00m          2.88  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-encode-locale                   \u001b[00m          1.05  pl526_6           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-error                           \u001b[00m       0.17027  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-eval-closure                    \u001b[00m          0.14  pl526h6bb024c_4   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-exception-class                 \u001b[00m          1.44  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-exporter                        \u001b[00m          5.72  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-exporter-tiny                   \u001b[00m      1.002001  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-extutils-makemaker              \u001b[00m          7.36  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-file-listing                    \u001b[00m          6.04  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-file-path                       \u001b[00m          2.16  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-file-slurp-tiny                 \u001b[00m         0.004  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-file-sort                       \u001b[00m          1.01  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-file-temp                       \u001b[00m        0.2304  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-file-which                      \u001b[00m          1.23  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-font-afm                        \u001b[00m          1.20  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-font-ttf                        \u001b[00m          1.06  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-gd                              \u001b[00m          2.71  pl526he860b03_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-getopt-long                     \u001b[00m          2.50  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-graph                           \u001b[00m        0.9704  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-graphviz                        \u001b[00m          2.24  pl526h734ff71_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-html-element-extended           \u001b[00m          1.18  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-html-entities-numbered          \u001b[00m          0.04  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-html-formatter                  \u001b[00m          2.16  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-html-parser                     \u001b[00m          3.72  pl526h6bb024c_5   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-html-tableextract               \u001b[00m          2.13  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-html-tagset                     \u001b[00m          3.20  pl526_3           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-html-tidy                       \u001b[00m          1.60  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-html-tree                       \u001b[00m          5.07  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-html-treebuilder-xpath          \u001b[00m          0.14  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-http-cookies                    \u001b[00m          6.04  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-http-daemon                     \u001b[00m          6.01  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-http-date                       \u001b[00m          6.02  pl526_3           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-http-message                    \u001b[00m          6.18  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-http-negotiate                  \u001b[00m          6.01  pl526_3           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-image-info                      \u001b[00m          1.38  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-image-size                      \u001b[00m         3.300  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-io-compress                     \u001b[00m         2.087  pl526he1b5a44_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-io-html                         \u001b[00m         1.001  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-io-sessiondata                  \u001b[00m          1.03  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-io-socket-ssl                   \u001b[00m         2.066  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-io-string                       \u001b[00m          1.08  pl526_3           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-io-stringy                      \u001b[00m         2.111  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-io-tty                          \u001b[00m          1.12  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-io-zlib                         \u001b[00m          1.10  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-ipc-run                         \u001b[00m    20180523.0  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-ipc-sharelite                   \u001b[00m          0.17  pl526h6bb024c_1   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-jcode                           \u001b[00m          2.07  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-json                            \u001b[00m          4.02  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-json-xs                         \u001b[00m          2.34  pl526h6bb024c_3   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-lib                             \u001b[00m          0.63  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-libwww-perl                     \u001b[00m          6.39  pl526_0           bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-libxml-perl                     \u001b[00m          0.08  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-list-moreutils                  \u001b[00m         0.428  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-list-moreutils-xs               \u001b[00m         0.428  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-lwp-mediatypes                  \u001b[00m          6.04  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-lwp-protocol-https              \u001b[00m          6.07  pl526_4           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-lwp-simple                      \u001b[00m          6.15  pl526h470a237_4   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-mailtools                       \u001b[00m          2.21  pl526_0           bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-math-cdf                        \u001b[00m           0.1  pl526h14c3975_5   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-math-derivative                 \u001b[00m          1.01  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-math-random                     \u001b[00m          0.72  pl526h14c3975_2   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-math-spline                     \u001b[00m          0.02  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-mime-base64                     \u001b[00m          3.15  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-mime-lite                       \u001b[00m         3.030  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-mime-tools                      \u001b[00m         5.508  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-mime-types                      \u001b[00m          2.17  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-mldbm                           \u001b[00m          2.05  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-module-implementation           \u001b[00m          0.09  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-module-runtime                  \u001b[00m         0.016  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-module-runtime-conflicts        \u001b[00m         0.003  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-moo                             \u001b[00m      2.003004  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-moose                           \u001b[00m        2.2011  pl526hf484d3e_1   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-mozilla-ca                      \u001b[00m      20180117  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-mro-compat                      \u001b[00m          0.13  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-net-http                        \u001b[00m          6.19  pl526_0           bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-net-ssleay                      \u001b[00m          1.74  0                 bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-ntlm                            \u001b[00m          1.09  pl526_4           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-ole-storage_lite                \u001b[00m          0.19  pl526_3           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-package-deprecationmanager      \u001b[00m          0.17  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-package-stash                   \u001b[00m          0.38  pl526hf484d3e_1   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-package-stash-xs                \u001b[00m          0.28  pl526hf484d3e_1   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-params-util                     \u001b[00m          1.07  pl526h6bb024c_4   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-parent                          \u001b[00m         0.236  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-parse-recdescent                \u001b[00m      1.967015  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-pathtools                       \u001b[00m          3.75  pl526h14c3975_1   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-pdf-api2                        \u001b[00m         2.035  pl526_0           bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-pod-escapes                     \u001b[00m          1.07  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-pod-usage                       \u001b[00m          1.69  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-postscript                      \u001b[00m          0.06  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-role-tiny                       \u001b[00m      2.000008  pl526_0           bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-scalar-list-utils               \u001b[00m          1.52  pl526h516909a_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-set-scalar                      \u001b[00m          1.29  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-soap-lite                       \u001b[00m          1.19  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-socket                          \u001b[00m         2.027  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-sort-naturally                  \u001b[00m          1.03  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-spreadsheet-parseexcel          \u001b[00m          0.65  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-spreadsheet-writeexcel          \u001b[00m          2.40  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-statistics-descriptive          \u001b[00m        3.0702  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-storable                        \u001b[00m          3.15  pl526h14c3975_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-sub-exporter                    \u001b[00m         0.987  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-sub-exporter-progressive        \u001b[00m      0.001013  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-sub-identify                    \u001b[00m          0.14  pl526h14c3975_0   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-sub-install                     \u001b[00m         0.928  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-sub-name                        \u001b[00m          0.21  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-sub-quote                       \u001b[00m      2.006003  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-sub-uplevel                     \u001b[00m        0.2800  pl526h14c3975_2   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-svg                             \u001b[00m          2.84  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-svg-graph                       \u001b[00m          0.02  pl526_3           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-task-weaken                     \u001b[00m          1.06  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-template-toolkit                \u001b[00m          2.26  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-test                            \u001b[00m          1.26  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-test-deep                       \u001b[00m         1.128  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-test-differences                \u001b[00m          0.67  pl526_0           bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-test-exception                  \u001b[00m          0.43  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-test-harness                    \u001b[00m          3.42  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-test-leaktrace                  \u001b[00m          0.16  pl526h14c3975_2   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-test-most                       \u001b[00m          0.35  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-test-requiresinternet           \u001b[00m          0.05  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-test-warn                       \u001b[00m          0.36  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-text-diff                       \u001b[00m          1.45  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-threaded                        \u001b[00m        5.32.1  hdfd78af_1        bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-tie-ixhash                      \u001b[00m          1.23  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-time-hires                      \u001b[00m        1.9760  pl526h14c3975_1   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-time-local                      \u001b[00m          1.28  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-timedate                        \u001b[00m          2.30  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-tree-dag_node                   \u001b[00m          1.31  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-try-tiny                        \u001b[00m          0.30  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-type-tiny                       \u001b[00m      1.004004  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-types-serialiser                \u001b[00m           1.0  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-unicode-map                     \u001b[00m         0.112  pl526h6bb024c_3   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-uri                             \u001b[00m          1.76  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-www-robotrules                  \u001b[00m          6.02  pl526_3           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-dom                         \u001b[00m          1.46  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-dom-xpath                   \u001b[00m          0.14  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-filter-buffertext           \u001b[00m          1.01  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-libxml                      \u001b[00m        2.0132  pl526h7ec2d77_1   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-libxslt                     \u001b[00m          1.94  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-namespacesupport            \u001b[00m          1.12  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-parser                      \u001b[00m          2.44  pl526h4e0c4b3_7   bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-regexp                      \u001b[00m          0.04  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-sax                         \u001b[00m          1.02  pl526_0           bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-sax-base                    \u001b[00m          1.09  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-sax-expat                   \u001b[00m          0.51  pl526_3           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-sax-writer                  \u001b[00m          0.57  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-simple                      \u001b[00m          2.25  pl526_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-twig                        \u001b[00m          3.52  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-writer                      \u001b[00m         0.625  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-xpath                       \u001b[00m          1.44  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xml-xpathengine                 \u001b[00m          0.14  pl526_2           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xsloader                        \u001b[00m          0.24  pl526_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-yaml                            \u001b[00m          1.29  pl526_0           bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pip                                  \u001b[00m        20.1.1  pyh9f0ad1d_0      conda-forge/noarch  \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pixman                               \u001b[00m        0.40.0  h36c2ea0_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pthread-stubs                        \u001b[00m           0.4  h36c2ea0_1001     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + python                               \u001b[00m        2.7.18  ha1903f6_2        pkgs/main/linux-64  \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + python_abi                           \u001b[00m           2.7  1_cp27mu          conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + readline                             \u001b[00m           8.1  h46c0cb4_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + setuptools                           \u001b[00m        44.0.0  py27_0            conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + sqlite                               \u001b[00m        3.37.0  h9cd32fc_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + t_coffee                             \u001b[00m        11.0.8  py27_2            bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + tidyp                                \u001b[00m          1.04  h779adbc_3        bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + tk                                   \u001b[00m        8.6.11  h27826a3_1        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + wheel                                \u001b[00m        0.37.1  pyhd8ed1ab_0      conda-forge/noarch  \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-kbproto                         \u001b[00m         1.0.7  h7f98852_1002     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-libice                          \u001b[00m        1.0.10  h7f98852_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-libsm                           \u001b[00m         1.2.3  hd9c2040_1000     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-libx11                          \u001b[00m         1.7.2  h7f98852_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-libxau                          \u001b[00m         1.0.9  h7f98852_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-libxdmcp                        \u001b[00m         1.1.3  h7f98852_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-libxext                         \u001b[00m         1.3.4  h7f98852_1        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-libxpm                          \u001b[00m        3.5.13  h7f98852_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-libxrender                      \u001b[00m        0.9.10  h7f98852_1003     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-libxt                           \u001b[00m         1.2.1  h7f98852_2        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-renderproto                     \u001b[00m        0.11.1  h7f98852_1002     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-xextproto                       \u001b[00m         7.3.0  h7f98852_1002     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xorg-xproto                          \u001b[00m        7.0.31  h7f98852_1007     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xz                                   \u001b[00m         5.2.5  h516909a_1        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + zlib                                 \u001b[00m        1.2.11  h36c2ea0_1013     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + zstd                                 \u001b[00m         1.4.9  ha95c52a_0        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\n",
      "  Summary:\n",
      "\n",
      "  Install: 291 packages\n",
      "\n",
      "  Total download: 0  B\n",
      "\n",
      "\n",
      "\n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Text starting with \"#\" is a comment as not a command\n",
    "# The below is the equivalent `conda` command which we are replacing with `mamba`\n",
    "# I use \"-q\" to print less information when running mamba\n",
    "# conda create -y -n mlst mlst\n",
    "mamba create -q -y -n mlst mlst"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84dc46e0-f86a-4ae0-9cad-0052bcdf1f1e",
   "metadata": {},
   "source": [
    "Once the tools are installed with `mamba` we used the regular `conda` command to run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fb2a7f9d-4eb9-4d76-839b-b43dbb3f2159",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR3028792.fasta\tsenterica\t15\taroC(2)\tdnaN(7)\themD(9)\thisD(9)\tpurE(5)\tsucA(9)\tthrA(12)\n",
      "\n",
      "[14:22:22] This is mlst 2.19.0 running on linux with Perl 5.026002\n",
      "[14:22:22] Checking mlst dependencies:\n",
      "[14:22:22] Found 'blastn' => /home/CSCScience.ca/apetkau/miniconda3/envs/mlst/bin/blastn\n",
      "[14:22:22] Found 'any2fasta' => /home/CSCScience.ca/apetkau/miniconda3/envs/mlst/bin/any2fasta\n",
      "[14:22:23] Found blastn: 2.12.0+ (002012)\n",
      "[14:22:23] Excluding 2 schemes: ecoli_2 abaumannii\n",
      "[14:22:24] Found exact allele match senterica.hemD-9\n",
      "[14:22:24] Found exact allele match senterica.hisD-9\n",
      "[14:22:24] Found exact allele match senterica.dnaN-7\n",
      "[14:22:24] Found exact allele match ecoli.gyrB-213\n",
      "[14:22:24] Found exact allele match senterica.aroC-2\n",
      "[14:22:24] Found exact allele match senterica.sucA-9\n",
      "[14:22:24] Found exact allele match senterica.purE-5\n",
      "[14:22:24] Found exact allele match senterica.thrA-12\n",
      "[14:22:25] You can use --label XXX to replace an ugly filename in the output.\n",
      "[14:22:25] Done.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n mlst mlst SRR3028792.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd61b11a-6e2d-4f24-904c-5ff18f83b2b1",
   "metadata": {},
   "source": [
    "## 6.3. Serotyping\n",
    "\n",
    "As this is a *Salmonella* genome, try running the [sistr](https://github.com/phac-nml/sistr_cmd) command on the genome to get the [serotype](https://en.wikipedia.org/wiki/Serotype)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a480fae1-a58c-4895-aa84-20b6c0ca524b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Package                         Version  Build                Channel                    Size\n",
      "\n",
      "  Install:\n",
      "\n",
      "\n",
      "\u001b[32m  + _libgcc_mutex          \u001b[00m           0.1  conda_forge          conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + _openmp_mutex          \u001b[00m           4.5  1_gnu                conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + blast                  \u001b[00m        2.12.0  pl5262h3289130_0     bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + blosc                  \u001b[00m        1.21.0  h9c3ff4c_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + bzip2                  \u001b[00m         1.0.8  h7f98852_4           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + c-ares                 \u001b[00m        1.18.1  h7f98852_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + ca-certificates        \u001b[00m     2021.10.8  ha878542_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + capnproto              \u001b[00m         0.9.1  h812cca2_4           conda-forge/linux-64       5 MB\n",
      "\u001b[32m  + curl                   \u001b[00m        7.81.0  h2574ce0_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + entrez-direct          \u001b[00m          16.2  he881be0_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + gsl                    \u001b[00m           2.6  he838d99_2           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + hdf5                   \u001b[00m        1.12.1  nompi_h2750804_103   conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + krb5                   \u001b[00m        1.19.2  hcc1bbae_3           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + ld_impl_linux-64       \u001b[00m        2.36.1  hea4e1c9_2           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libblas                \u001b[00m         3.9.0  13_linux64_openblas  conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libcblas               \u001b[00m         3.9.0  13_linux64_openblas  conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libcurl                \u001b[00m        7.81.0  h2574ce0_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libedit                \u001b[00m  3.1.20191231  he28a2e2_2           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libev                  \u001b[00m          4.33  h516909a_1           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libffi                 \u001b[00m         3.4.2  h7f98852_5           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libgcc-ng              \u001b[00m        11.2.0  h1d223b6_11          conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libgfortran-ng         \u001b[00m        11.2.0  h69a702a_11          conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libgfortran5           \u001b[00m        11.2.0  h5c6108e_11          conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libgomp                \u001b[00m        11.2.0  h1d223b6_11          conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + liblapack              \u001b[00m         3.9.0  13_linux64_openblas  conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libnghttp2             \u001b[00m        1.43.0  h812cca2_1           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libnsl                 \u001b[00m         2.0.0  h7f98852_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libopenblas            \u001b[00m        0.3.18  pthreads_h8fe5266_0  conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libssh2                \u001b[00m        1.10.0  ha56f1ee_2           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libstdcxx-ng           \u001b[00m        11.2.0  he4da1e4_11          conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + libzlib                \u001b[00m        1.2.11  h36c2ea0_1013        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + lzo                    \u001b[00m          2.10  h516909a_1000        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + mafft                  \u001b[00m         7.490  h779adbc_0           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + mash                   \u001b[00m           2.3  he348c14_1           bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + mock                   \u001b[00m         4.0.3  py38h578d9bd_2       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + ncurses                \u001b[00m           6.3  h9c3ff4c_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + nomkl                  \u001b[00m           1.0  h5ca1d4c_0           conda-forge/noarch  \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + numexpr                \u001b[00m         2.8.0  py38h6045d29_100     conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + numpy                  \u001b[00m        1.22.0  py38h6ae9a64_1       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + openssl                \u001b[00m        1.1.1l  h7f98852_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pandas                 \u001b[00m         1.0.5  py38hcb8c335_0       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pcre                   \u001b[00m          8.45  h9c3ff4c_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl                   \u001b[00m        5.26.2  h36c2ea0_1008        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-archive-tar       \u001b[00m          2.32  pl526_0              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-carp              \u001b[00m          1.38  pl526_3              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-common-sense      \u001b[00m          3.74  pl526_2              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-compress-raw-bzip2\u001b[00m         2.087  pl526he1b5a44_0      bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-compress-raw-zlib \u001b[00m         2.087  pl526hc9558a2_0      bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-exporter          \u001b[00m          5.72  pl526_1              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-exporter-tiny     \u001b[00m      1.002001  pl526_0              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-extutils-makemaker\u001b[00m          7.36  pl526_1              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-io-compress       \u001b[00m         2.087  pl526he1b5a44_0      bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-io-zlib           \u001b[00m          1.10  pl526_2              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-json              \u001b[00m          4.02  pl526_0              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-json-xs           \u001b[00m          2.34  pl526h6bb024c_3      bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-list-moreutils    \u001b[00m         0.428  pl526_1              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-list-moreutils-xs \u001b[00m         0.428  pl526_0              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-pathtools         \u001b[00m          3.75  pl526h14c3975_1      bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-scalar-list-utils \u001b[00m          1.52  pl526h516909a_0      bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-types-serialiser  \u001b[00m           1.0  pl526_2              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + perl-xsloader          \u001b[00m          0.24  pl526_0              bioconda/linux-64   \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pip                    \u001b[00m        21.3.1  pyhd8ed1ab_0         conda-forge/noarch  \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pycurl                 \u001b[00m        7.44.1  py38h996a351_1       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pytables               \u001b[00m         3.6.1  py38hdb04529_5       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + python                 \u001b[00m        3.8.12  hb7a2778_2_cpython   conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + python-dateutil        \u001b[00m         2.8.2  pyhd8ed1ab_0         conda-forge/noarch  \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + python_abi             \u001b[00m           3.8  2_cp38               conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + pytz                   \u001b[00m        2021.3  pyhd8ed1ab_0         conda-forge/noarch  \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + readline               \u001b[00m           8.1  h46c0cb4_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + scipy                  \u001b[00m         1.7.3  py38h56a6a73_0       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + setuptools             \u001b[00m        59.8.0  py38h578d9bd_0       conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + sistr_cmd              \u001b[00m         1.1.1  pyh864c0ab_2         bioconda/noarch     \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + six                    \u001b[00m        1.16.0  pyh6c4a22f_0         conda-forge/noarch  \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + sqlite                 \u001b[00m        3.37.0  h9cd32fc_0           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + tk                     \u001b[00m        8.6.11  h27826a3_1           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + wheel                  \u001b[00m        0.37.1  pyhd8ed1ab_0         conda-forge/noarch  \u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + xz                     \u001b[00m         5.2.5  h516909a_1           conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\u001b[32m  + zlib                   \u001b[00m        1.2.11  h36c2ea0_1013        conda-forge/linux-64\u001b[32m     Cached\u001b[00m\n",
      "\n",
      "  Summary:\n",
      "\n",
      "  Install: 78 packages\n",
      "\n",
      "  Total download: 5 MB\n",
      "\n",
      "\n",
      "\n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "mamba create -q -y -n sistr sistr_cmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b216fe9b-d95c-4c7d-ab94-a903c38d943f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sistr_cmd [-h] [-i fasta_path genome_name] [-f OUTPUT_FORMAT]\n",
      "                 [-o OUTPUT_PREDICTION] [-M] [-p CGMLST_PROFILES]\n",
      "                 [-n NOVEL_ALLELES] [-a ALLELES_OUTPUT] [-T TMP_DIR] [-K]\n",
      "                 [--use-full-cgmlst-db] [--no-cgmlst] [-m] [--qc] [-t THREADS]\n",
      "                 [-v] [-V]\n",
      "                 [F [F ...]]\n",
      "\n",
      "SISTR (Salmonella In Silico Typing Resource) Command-line Tool\n",
      "==============================================================\n",
      "Serovar predictions from whole-genome sequence assemblies by determination of antigen gene and cgMLST gene alleles using BLAST.\n",
      "\n",
      "Note about using the \"--use-full-cgmlst-db\" flag:\n",
      "    The \"centroid\" allele database is ~10% the size of the full set so analysis is much quicker with the \"centroid\" vs \"full\" set of alleles. Results between 2 cgMLST allele sets should not differ.\n",
      "\n",
      "If you find this program useful in your research, please cite as:\n",
      "\n",
      "The Salmonella In Silico Typing Resource (SISTR): an open web-accessible tool for rapidly typing and subtyping draft Salmonella genome assemblies.\n",
      "Catherine Yoshida, Peter Kruczkiewicz, Chad R. Laing, Erika J. Lingohr, Victor P.J. Gannon, John H.E. Nash, Eduardo N. Taboada.\n",
      "PLoS ONE 11(1): e0147101. doi: 10.1371/journal.pone.0147101\n",
      "\n",
      "positional arguments:\n",
      "  F                     Input genome FASTA file\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -i fasta_path genome_name, --input-fasta-genome-name fasta_path genome_name\n",
      "                        fasta file path to genome name pair\n",
      "  -f OUTPUT_FORMAT, --output-format OUTPUT_FORMAT\n",
      "                        Output format (json, csv, pickle)\n",
      "  -o OUTPUT_PREDICTION, --output-prediction OUTPUT_PREDICTION\n",
      "                        SISTR serovar prediction output path\n",
      "  -M, --more-results    Output more detailed results (-M) and all antigen\n",
      "                        search blastn results (-MM)\n",
      "  -p CGMLST_PROFILES, --cgmlst-profiles CGMLST_PROFILES\n",
      "                        Output CSV file destination for cgMLST allelic\n",
      "                        profiles\n",
      "  -n NOVEL_ALLELES, --novel-alleles NOVEL_ALLELES\n",
      "                        Output FASTA file destination of novel cgMLST alleles\n",
      "                        from input genomes\n",
      "  -a ALLELES_OUTPUT, --alleles-output ALLELES_OUTPUT\n",
      "                        Output path of allele sequences and info to JSON\n",
      "  -T TMP_DIR, --tmp-dir TMP_DIR\n",
      "                        Base temporary working directory for intermediate\n",
      "                        analysis files.\n",
      "  -K, --keep-tmp        Keep temporary analysis files.\n",
      "  --use-full-cgmlst-db  Use the full set of cgMLST alleles which can include\n",
      "                        highly similar alleles. By default the smaller\n",
      "                        \"centroid\" alleles or representative alleles are used\n",
      "                        for each marker.\n",
      "  --no-cgmlst           Do not run cgMLST serovar prediction\n",
      "  -m, --run-mash        Determine Mash MinHash genomic distances to Salmonella\n",
      "                        genomes with trusted serovar designations. Mash binary\n",
      "                        must be in accessible via $PATH (e.g. /usr/bin).\n",
      "  --qc                  Perform basic QC to provide level of confidence in\n",
      "                        serovar prediction results.\n",
      "  -t THREADS, --threads THREADS\n",
      "                        Number of parallel threads to run sistr_cmd analysis.\n",
      "  -v, --verbose         Logging verbosity level (-v == show warnings; -vvv ==\n",
      "                        show debug info)\n",
      "  -V, --version         show program's version number and exit\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n sistr sistr --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7c680e9b-9db6-4543-9cb0-a3b9b5144cec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "conda run -n sistr sistr -f csv -o SRR3028792.sistr.csv SRR3028792.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8556e7e2-9dfe-404b-9649-0fcbee70e76c",
   "metadata": {},
   "source": [
    "The output is saved to the file `SRR3028792.sistr.csv`. We can view it with `cat`. Also, since it's a **csv** file, you can download it and open it in any spreadsheet software (e.g., Excel)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e95e1269-969f-424d-9d98-ed1f4c6a6615",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cgmlst_ST,cgmlst_distance,cgmlst_found_loci,cgmlst_genome_match,cgmlst_matching_alleles,cgmlst_subspecies,fasta_filepath,genome,h1,h2,o_antigen,serogroup,serovar,serovar_antigen,serovar_cgmlst\n",
      "3331578236,0.018181818181818188,330,2017-MER-0640,324,enterica,/home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/SRR3028792.fasta,SRR3028792,r,\"1,2\",\"1,4,[5],12\",B,Heidelberg,Heidelberg|Bradford,Heidelberg\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cat SRR3028792.sistr.csv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Calysto Bash",
   "language": "bash",
   "name": "calysto_bash"
  },
  "language_info": {
   "file_extension": ".sh",
   "help_links": [
    {
     "text": "MetaKernel Magics",
     "url": "https://metakernel.readthedocs.io/en/latest/source/README.html"
    }
   ],
   "mimetype": "text/x-sh",
   "name": "bash",
   "version": "0.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
